// @okendo/okendo-widgets-vue/dist/okendo-widgets-vue.esm.js
import axios from 'axios';

const productIdRegex = /^[0-9]*$/;
const apiBaseUrl = 'https://api.okendo.io/v1';
let subscriberId = undefined;
const aggregateCache = {};
var OkendoAPI = {
  get storeBaseUrl() {
    if (subscriberId) {
      return `${apiBaseUrl}/stores/${subscriberId}`;
    } else {
      console.error("Okendo subscriberId not defined: Please set your Okendo subscriberId by calling configureOkendo({ subscriberId: 'your-subscriber-id-here' })");
    }
  },

  transformProductId(productId) {
    if (productId.startsWith('shopify-')) {
      return productId;
    } else {
      // Supports Shopify's GraphQL Id syntax e.g. gid://shopify/Product/10079785100
      return `shopify-${productIdRegex.test(productId) ? productId : atob(productId).split('/').slice(-1)[0]}`;
    }
  },

  setSubscriberId(newSubscriberId) {
    subscriberId = newSubscriberId;
  },

  apiBaseUrl() {
    return apiBaseUrl;
  },

  subscriberId() {
    return subscriberId;
  },

  writeReviewUrl(productId) {
    return `https://reviews.okendo.io/stores/${subscriberId}/review?productId=${productId}`;
  },

  askQuestionUrl(productId) {
    return `https://reviews.okendo.io/stores/${subscriberId}/question?productId=${productId}`;
  },

  getMore(nextUrl) {
    return axios.get(`${apiBaseUrl}${nextUrl}`);
  },

  postHelpfulVoteReview(reviewId, voteType) {
    return axios.post(`${this.storeBaseUrl}/reviews/${reviewId}/vote`, {
      voteType
    });
  },

  getReview(reviewId) {
    return axios.get(`${this.storeBaseUrl}/reviews/${reviewId}`);
  },

  getReviewsProduct(productId, limit = 5, orderBy = 'date desc') {
    return axios.get(`${this.storeBaseUrl}/products/${productId}/reviews?limit=${limit}&orderBy=${orderBy}`);
  },

  getReviewsAll(limit = 5, orderBy = 'date desc') {
    return axios.get(`${this.storeBaseUrl}/reviews?limit=${limit}&orderBy=${orderBy}`);
  },

  getReviewAggregateProduct(productId) {
    return aggregateCache[productId] || (aggregateCache[productId] = axios.get(`${this.storeBaseUrl}/products/${productId}/review_aggregate`));
  },

  getReviewAggregateAll() {
    return axios.get(`${this.storeBaseUrl}/review_aggregate`);
  },

  getMediaProduct(productId, limit = 6, orderBy = 'date desc') {
    return axios.get(`${this.storeBaseUrl}/products/${productId}/review_media?limit=${limit}&orderBy=${orderBy}`);
  },

  getMediaAll(limit = 6, orderBy = 'date desc') {
    return axios.get(`${this.storeBaseUrl}/review_media?limit=${limit}&orderBy=${orderBy}`);
  },

  getQuestionsProduct(productId, limit = 5, orderBy = 'date desc') {
    return axios.get(`${this.storeBaseUrl}/products/${productId}/questions?limit=${limit}&orderBy=${orderBy}`);
  },

  getQuestionAggregateProduct(productId) {
    return axios.get(`${this.storeBaseUrl}/products/${productId}/question_aggregate`);
  },

  postHelpfulVoteQuestion(questionId, voteType) {
    return axios.post(`${this.storeBaseUrl}/questions/${questionId}/vote`, {
      voteType
    });
  },

  postHelpfulVoteAnswer(questionId, answerId, voteType) {
    return axios.post(`${this.storeBaseUrl}/questions/${questionId}/answers/${answerId}/vote`, {
      voteType
    });
  }

};

//
//
//
//
//
//
//
//
//
//
var script = {
  props: {
    rating: {
      required: true
    },
    accessibleText: {
      type: Boolean,
      default: true,
      required: false
    },
    size: {
      type: String,
      default: 'standard',
      required: false
    }
  }
};

function normalizeComponent(template, style, script, scopeId, isFunctionalTemplate, moduleIdentifier /* server only */, shadowMode, createInjector, createInjectorSSR, createInjectorShadow) {
    if (typeof shadowMode !== 'boolean') {
        createInjectorSSR = createInjector;
        createInjector = shadowMode;
        shadowMode = false;
    }
    // Vue.extend constructor export interop.
    const options = typeof script === 'function' ? script.options : script;
    // render functions
    if (template && template.render) {
        options.render = template.render;
        options.staticRenderFns = template.staticRenderFns;
        options._compiled = true;
        // functional template
        if (isFunctionalTemplate) {
            options.functional = true;
        }
    }
    // scopedId
    if (scopeId) {
        options._scopeId = scopeId;
    }
    let hook;
    if (moduleIdentifier) {
        // server build
        hook = function (context) {
            // 2.3 injection
            context =
                context || // cached call
                    (this.$vnode && this.$vnode.ssrContext) || // stateful
                    (this.parent && this.parent.$vnode && this.parent.$vnode.ssrContext); // functional
            // 2.2 with runInNewContext: true
            if (!context && typeof __VUE_SSR_CONTEXT__ !== 'undefined') {
                context = __VUE_SSR_CONTEXT__;
            }
            // inject component styles
            if (style) {
                style.call(this, createInjectorSSR(context));
            }
            // register component module identifier for async chunk inference
            if (context && context._registeredComponents) {
                context._registeredComponents.add(moduleIdentifier);
            }
        };
        // used by ssr in case component is cached and beforeCreate
        // never gets called
        options._ssrRegister = hook;
    }
    else if (style) {
        hook = shadowMode
            ? function (context) {
                style.call(this, createInjectorShadow(context, this.$root.$options.shadowRoot));
            }
            : function (context) {
                style.call(this, createInjector(context));
            };
    }
    if (hook) {
        if (options.functional) {
            // register for functional component in vue file
            const originalRender = options.render;
            options.render = function renderWithStyleInjection(h, context) {
                hook.call(context);
                return originalRender(h, context);
            };
        }
        else {
            // inject component registration as beforeCreate hook
            const existing = options.beforeCreate;
            options.beforeCreate = existing ? [].concat(existing, hook) : [hook];
        }
    }
    return script;
}

/* script */
const __vue_script__ = script;

/* template */
var __vue_render__ = function () {var _vm=this;var _h=_vm.$createElement;var _c=_vm._self._c||_h;return _c('span',{staticClass:"okeReviews-starRating",class:{ 'okeReviews-starRating--small': _vm.size === 'small' }},[(_vm.accessibleText)?_c('span',{staticClass:"okeReviews-a11yText"},[_vm._v("Rated "+_vm._s(_vm.rating)+" out of 5")]):_vm._e(),_vm._v(" "),_c('span',{staticClass:"okeReviews-starRating-indicator",attrs:{"role":"presentation"}},[_c('span',{staticClass:"okeReviews-starRating-indicator-layer"}),_vm._v(" "),_c('span',{staticClass:"okeReviews-starRating-indicator-layer okeReviews-starRating-indicator-layer--foreground",style:({ width: _vm.rating * 20 + '%' })})])])};
var __vue_staticRenderFns__ = [];

  /* style */
  const __vue_inject_styles__ = undefined;
  /* scoped */
  const __vue_scope_id__ = undefined;
  /* module identifier */
  const __vue_module_identifier__ = undefined;
  /* functional template */
  const __vue_is_functional_template__ = false;
  /* style inject */
  
  /* style inject SSR */
  
  /* style inject shadow dom */
  

  
  var OkendoStars = normalizeComponent(
    { render: __vue_render__, staticRenderFns: __vue_staticRenderFns__ },
    __vue_inject_styles__,
    __vue_script__,
    __vue_scope_id__,
    __vue_is_functional_template__,
    __vue_module_identifier__,
    false,
    undefined,
    undefined,
    undefined
  );

//
var script$1 = {
  props: {
    productId: {
      type: String,
      default: '',
      required: true
    },
    // Hide stars when there are no reviews for the product
    hideIfEmpty: {
      type: Boolean,
      default: true,
      required: false
    },
    // Id of element to scroll to when stars are clicked (defaults to reviews widget)
    scrollTargetId: {
      type: String,
      required: false
    }
  },
  components: {
    OkendoStars
  },

  data() {
    return {
      ratingAndReviewCount: 0,
      averageRating: '0',
      isLoading: true,
      isError: false,
      scrollTarget: undefined
    };
  },

  mounted() {
    const productId = OkendoAPI.transformProductId(this.productId);
    this.scrollTarget = this.scrollTargetId ? document.getElementById(this.scrollTargetId) : document.getElementsByClassName('js-okeReviews-reviewsWidget')[0];

    if (productId) {
      OkendoAPI.getReviewAggregateProduct(productId).then(response => {
        if (response && response.data && response.data.reviewAggregate) {
          const reviewAggregate = response.data.reviewAggregate;
          this.ratingAndReviewCount = reviewAggregate.ratingAndReviewCount;
          this.averageRating = reviewAggregate.ratingAndReviewCount ? (reviewAggregate.ratingAndReviewValuesTotal / reviewAggregate.ratingAndReviewCount).toFixed(1) : '0';
          this.isLoading = false;
        } else {
          this.isError = true;
        }
      }).catch(() => this.isError = true);
    }
  },

  methods: {
    scrollToReviews() {
      if (this.scrollTarget) {
        this.scrollTarget.scrollIntoView({
          behavior: 'smooth'
        });
      }
    }

  }
};

const isOldIE = typeof navigator !== 'undefined' &&
    /msie [6-9]\\b/.test(navigator.userAgent.toLowerCase());
function createInjector(context) {
    return (id, style) => addStyle(id, style);
}
let HEAD;
const styles = {};
function addStyle(id, css) {
    const group = isOldIE ? css.media || 'default' : id;
    const style = styles[group] || (styles[group] = { ids: new Set(), styles: [] });
    if (!style.ids.has(id)) {
        style.ids.add(id);
        let code = css.source;
        if (css.map) {
            // https://developer.chrome.com/devtools/docs/javascript-debugging
            // this makes source maps inside style tags work properly in Chrome
            code += '\n/*# sourceURL=' + css.map.sources[0] + ' */';
            // http://stackoverflow.com/a/26603875
            code +=
                '\n/*# sourceMappingURL=data:application/json;base64,' +
                    btoa(unescape(encodeURIComponent(JSON.stringify(css.map)))) +
                    ' */';
        }
        if (!style.element) {
            style.element = document.createElement('style');
            style.element.type = 'text/css';
            if (css.media)
                style.element.setAttribute('media', css.media);
            if (HEAD === undefined) {
                HEAD = document.head || document.getElementsByTagName('head')[0];
            }
            HEAD.appendChild(style.element);
        }
        if ('styleSheet' in style.element) {
            style.styles.push(code);
            style.element.styleSheet.cssText = style.styles
                .filter(Boolean)
                .join('\n');
        }
        else {
            const index = style.ids.size - 1;
            const textNode = document.createTextNode(code);
            const nodes = style.element.childNodes;
            if (nodes[index])
                style.element.removeChild(nodes[index]);
            if (nodes.length)
                style.element.insertBefore(textNode, nodes[index]);
            else
                style.element.appendChild(textNode);
        }
    }
}

/* script */
const __vue_script__$1 = script$1;

/* template */
var __vue_render__$1 = function () {var _vm=this;var _h=_vm.$createElement;var _c=_vm._self._c||_h;return (!_vm.isError && (_vm.isLoading || _vm.ratingAndReviewCount || !_vm.hideIfEmpty))?_c('div',{staticClass:"okeReviews okeReviews--theme",on:{"click":function($event){return _vm.scrollToReviews()}}},[_c('div',{staticClass:"okeReviews-reviewsSummary",class:{
            'is-okeReviews-empty': _vm.ratingAndReviewCount === 0,
            'is-okeReviews-loading': _vm.isLoading,
            'is-okeReviews-clickable': _vm.scrollTarget
        }},[_c('div',{staticClass:"okeReviews-reviewsSummary-starRating"},[_c('okendo-stars',{attrs:{"rating":_vm.averageRating,"size":"small"}})],1),_vm._v(" "),_c('div',{staticClass:"okeReviews-reviewsSummary-ratingCount"},[_vm._v("\n            "+_vm._s(_vm.ratingAndReviewCount || '0')+" "+_vm._s(_vm.ratingAndReviewCount === 1 ? 'Review' : 'Reviews')+"\n        ")])])]):_vm._e()};
var __vue_staticRenderFns__$1 = [];

  /* style */
  const __vue_inject_styles__$1 = function (inject) {
    if (!inject) return
    inject("data-v-5abc7252_0", { source: ".okeReviews .okeReviews-reviewsSummary-ratingCount,.okeReviews .okeReviews-starRating-indicator-layer--foreground{transition:opacity .5s ease}.okeReviews .okeReviews-reviewsSummary.is-okeReviews-loading{position:relative}.okeReviews .okeReviews-reviewsSummary.is-okeReviews-loading .okeReviews-reviewsSummary-ratingCount,.okeReviews .okeReviews-reviewsSummary.is-okeReviews-loading .okeReviews-starRating-indicator-layer--foreground{opacity:0}.okeReviews .okeReviews-reviewsSummary.is-okeReviews-loading .okeReviews-starRating-indicator::before{animation:loading 1.2s linear infinite;background:linear-gradient(to right,rgba(255,255,255,0) 0,rgba(255,255,255,.6) 100%);background-size:200% 100%;content:\"\";display:block;height:100%;position:absolute;width:100%;z-index:1}@keyframes loading{0%{background-position:200%}100%{background-position:0}}", map: undefined, media: undefined });

  };
  /* scoped */
  const __vue_scope_id__$1 = undefined;
  /* module identifier */
  const __vue_module_identifier__$1 = undefined;
  /* functional template */
  const __vue_is_functional_template__$1 = false;
  /* style inject SSR */
  
  /* style inject shadow dom */
  

  
  var OkendoStarRating = normalizeComponent(
    { render: __vue_render__$1, staticRenderFns: __vue_staticRenderFns__$1 },
    __vue_inject_styles__$1,
    __vue_script__$1,
    __vue_scope_id__$1,
    __vue_is_functional_template__$1,
    __vue_module_identifier__$1,
    false,
    createInjector,
    undefined,
    undefined
  );

//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
var script$2 = {
  props: {
    attributes: {
      type: Array,
      required: true
    }
  },
  methods: {
    isListAttribute(attribute) {
      return Array.isArray(attribute.value);
    }

  }
};

/* script */
const __vue_script__$2 = script$2;

/* template */
var __vue_render__$2 = function () {var _vm=this;var _h=_vm.$createElement;var _c=_vm._self._c||_h;return (_vm.attributes)?_c('div',{staticClass:"okeReviews-review-reviewer-attributes"},_vm._l((_vm.attributes),function(attribute){return _c('div',{key:attribute.title,staticClass:"okeReviews-review-reviewer-attributes-item"},[(attribute.value && _vm.isListAttribute(attribute))?_c('div',{staticClass:"okeReviews-sectionedList"},[_c('div',{staticClass:"okeReviews-sectionedList-heading"},[_vm._v(_vm._s(attribute.title))]),_vm._v(" "),_c('ul',{staticClass:"okeReviews-sectionedList-list"},_vm._l((attribute.value),function(value){return _c('li',{key:value,staticClass:"okeReviews-sectionedList-list-item"},[_vm._v(_vm._s(value))])}),0)]):_vm._e(),_vm._v(" "),(attribute.value && !_vm.isListAttribute(attribute))?_c('div',{staticClass:"okeReviews-review-attribute"},[_c('div',{staticClass:"okeReviews-review-attribute-label"},[_vm._v(_vm._s(attribute.title))]),_vm._v(" "),_c('div',{staticClass:"okeReviews-review-attribute-value"},[_vm._v(_vm._s(attribute.value))])]):_vm._e()])}),0):_vm._e()};
var __vue_staticRenderFns__$2 = [];

  /* style */
  const __vue_inject_styles__$2 = undefined;
  /* scoped */
  const __vue_scope_id__$2 = undefined;
  /* module identifier */
  const __vue_module_identifier__$2 = undefined;
  /* functional template */
  const __vue_is_functional_template__$2 = false;
  /* style inject */
  
  /* style inject SSR */
  
  /* style inject shadow dom */
  

  
  var OkendoCustomerAttributes = normalizeComponent(
    { render: __vue_render__$2, staticRenderFns: __vue_staticRenderFns__$2 },
    __vue_inject_styles__$2,
    __vue_script__$2,
    __vue_scope_id__$2,
    __vue_is_functional_template__$2,
    __vue_module_identifier__$2,
    false,
    undefined,
    undefined,
    undefined
  );

//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
var script$3 = {
  props: {
    attributes: {
      type: Array,
      required: true
    }
  },
  methods: {
    isListAttribute(attribute) {
      return Array.isArray(attribute.value);
    }

  }
};

/* script */
const __vue_script__$3 = script$3;

/* template */
var __vue_render__$3 = function () {var _vm=this;var _h=_vm.$createElement;var _c=_vm._self._c||_h;return (_vm.attributes)?_c('div',{staticClass:"okeReviews-review-arguments"},_vm._l((_vm.attributes),function(attribute){return _c('div',{key:attribute.title,staticClass:"okeReviews-review-arguments-item"},[(attribute.value && _vm.isListAttribute(attribute))?_c('div',{staticClass:"okeReviews-sectionedList"},[_c('div',{staticClass:"okeReviews-sectionedList-heading"},[_vm._v(_vm._s(attribute.title))]),_vm._v(" "),_c('ul',{staticClass:"okeReviews-sectionedList-list"},_vm._l((attribute.value),function(value){return _c('li',{key:value,staticClass:"okeReviews-sectionedList-list-item"},[_vm._v(_vm._s(value))])}),0)]):_vm._e(),_vm._v(" "),(attribute.value && !_vm.isListAttribute(attribute))?_c('div',{staticClass:"okeReviews-review-attribute"},[_c('div',{staticClass:"okeReviews-review-attribute-label"},[_vm._v(_vm._s(attribute.title))]),_vm._v(" "),_c('div',{staticClass:"okeReviews-review-attribute-value"},[_vm._v(_vm._s(attribute.value))])]):_vm._e()])}),0):_vm._e()};
var __vue_staticRenderFns__$3 = [];

  /* style */
  const __vue_inject_styles__$3 = undefined;
  /* scoped */
  const __vue_scope_id__$3 = undefined;
  /* module identifier */
  const __vue_module_identifier__$3 = undefined;
  /* functional template */
  const __vue_is_functional_template__$3 = false;
  /* style inject */
  
  /* style inject SSR */
  
  /* style inject shadow dom */
  

  
  var OkendoProductAttributes = normalizeComponent(
    { render: __vue_render__$3, staticRenderFns: __vue_staticRenderFns__$3 },
    __vue_inject_styles__$3,
    __vue_script__$3,
    __vue_scope_id__$3,
    __vue_is_functional_template__$3,
    __vue_module_identifier__$3,
    false,
    undefined,
    undefined,
    undefined
  );

//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
var script$4 = {
  props: {
    reviewer: {
      type: Object,
      required: true
    }
  },
  computed: {
    reviewerInitials() {
      return (this.reviewer.displayName || '').split(' ').slice(0, 2).map(part => part.substr(0, 1).toLocaleUpperCase()).join('');
    },

    reviewerLocation() {
      if (this.reviewer.location && this.reviewer.location.zoneCode) {
        return `${this.reviewer.location.zoneCode}, ${this.reviewer.location.country.name}`;
      } else if (this.reviewer.country) {
        return this.reviewer.country.name;
      }
    }

  }
};

/* script */
const __vue_script__$4 = script$4;

/* template */
var __vue_render__$4 = function () {var _vm=this;var _h=_vm.$createElement;var _c=_vm._self._c||_h;return _c('div',{staticClass:"okeReviews-review-reviewer-profile"},[_c('div',{staticClass:"okeReviews-review-reviewer-profile-avatar"},[_c('div',{staticClass:"okeReviews-avatar okeReviews-avatar--reviewer",class:_vm.reviewer.socialConnection ? ("okeReviews-avatar--" + (_vm.reviewer.socialConnection)) : ''},[_c('div',{staticClass:"okeReviews-avatar-placeholder",attrs:{"aria-hidden":"true"}},[_vm._v("\n                "+_vm._s(_vm.reviewerInitials)+"\n            ")]),_vm._v(" "),(_vm.reviewer.avatarUrl)?_c('img',{staticClass:"okeReviews-avatar-image",attrs:{"src":_vm.reviewer.avatarUrl,"alt":""}}):_vm._e()])]),_vm._v(" "),_c('div',{staticClass:"okeReviews-review-reviewer-profile-details"},[_c('div',{staticClass:"okeReviews-review-reviewer-profile-details-inner"},[_c('span',{staticClass:"okeReviews-review-reviewer-profile-name",attrs:{"aria-hidden":"true"}},[_vm._v("\n                "+_vm._s(_vm.reviewer.displayName)+"\n            ")]),_vm._v(" "),(_vm.reviewer.country)?[(_vm.reviewer.location && _vm.reviewer.location.zoneCode)?_c('span',{staticClass:"okeReviews-review-reviewer-profile-country-zone",attrs:{"aria-hidden":"true"}},[_vm._v("\n                    "+_vm._s(_vm.reviewerLocation)+"\n                ")]):_c('span',{staticClass:"okeReviews-review-reviewer-profile-country okeReviews-flag",class:("okeReviews-flag--" + (_vm.reviewer.country.code.toLowerCase())),attrs:{"title":_vm.reviewerLocation,"aria-hidden":"true"}}),_vm._v(" "),_c('span',{staticClass:"okeReviews-a11yText"},[_vm._v("\n                    "+_vm._s(("Reviewed by " + (_vm.reviewer.displayName) + ", from " + _vm.reviewerLocation))+"\n                ")])]:[_c('span',{staticClass:"okeReviews-a11yText"},[_vm._v(_vm._s(("Reviewed by " + (_vm.reviewer.displayName))))])]],2),_vm._v(" "),(_vm.reviewer.isVerified)?_c('div',{staticClass:"okeReviews-review-reviewer-profile-status okeReviews-review-reviewer-profile-status--verified"},[_vm._v("\n            Verified Buyer\n        ")]):_vm._e()])])};
var __vue_staticRenderFns__$4 = [];

  /* style */
  const __vue_inject_styles__$4 = function (inject) {
    if (!inject) return
    inject("data-v-49c2eb92_0", { source: ".okeReviews .okeReviews-review-reviewer-profile-country-zone[data-v-49c2eb92]{margin-bottom:0}", map: undefined, media: undefined });

  };
  /* scoped */
  const __vue_scope_id__$4 = "data-v-49c2eb92";
  /* module identifier */
  const __vue_module_identifier__$4 = undefined;
  /* functional template */
  const __vue_is_functional_template__$4 = false;
  /* style inject SSR */
  
  /* style inject shadow dom */
  

  
  var OkendoReviewerProfile = normalizeComponent(
    { render: __vue_render__$4, staticRenderFns: __vue_staticRenderFns__$4 },
    __vue_inject_styles__$4,
    __vue_script__$4,
    __vue_scope_id__$4,
    __vue_is_functional_template__$4,
    __vue_module_identifier__$4,
    false,
    createInjector,
    undefined,
    undefined
  );

//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
var script$5 = {
  props: {
    attributes: {
      type: Array,
      required: true
    }
  },
  methods: {
    percentageValue(index) {
      const attribute = this.attributes[index];
      return attribute.type === 'range' ? this.scaleBetween(attribute.value, 0, 100, 1, 5) : this.scaleBetween(attribute.value, 0, 100, -1, 1);
    },

    scaledValue(index) {
      const attribute = this.attributes[index];
      return attribute.type === 'range' ? attribute.value : this.scaleBetween(attribute.value, -2, 2, -1, 1);
    },

    scaleBetween(unscaledNum, minAllowed, maxAllowed, unscaledNumMin, unscaledNumMax) {
      return (maxAllowed - minAllowed) * (unscaledNum - unscaledNumMin) / (unscaledNumMax - unscaledNumMin) + minAllowed;
    }

  }
};

/* script */
const __vue_script__$5 = script$5;

/* template */
var __vue_render__$5 = function () {var _vm=this;var _h=_vm.$createElement;var _c=_vm._self._c||_h;return _c('tbody',_vm._l((_vm.attributes),function(attribute,index){return _c('tr',{key:attribute.title,staticClass:"okeReviews-attributeRatingsTable-row"},[_c('th',{staticClass:"okeReviews-attributeRatingsTable-row-label"},[_vm._v("\n            "+_vm._s(attribute.title)+"\n        ")]),_vm._v(" "),_c('td',{staticClass:"okeReviews-attributeRatingsTable-row-value"},[(attribute.type === 'range')?_c('div',{staticClass:"or-rg",class:{ 'has-lbl': attribute.minLabel || attribute.maxLabel }},[_c('span',{staticClass:"okeReviews-a11yText"},[_vm._v("Rated "+_vm._s(_vm.scaledValue(index))+" out of 5")]),_vm._v(" "),_c('div',{staticClass:"or-rg-bar"},[_c('div',{staticClass:"or-rg-bar-bg",style:({ width: _vm.percentageValue(index) + '%'})}),_vm._v(" "),_c('div',{staticClass:"or-rg-bar-seg"}),_vm._v(" "),_c('div',{staticClass:"or-rg-bar-seg"}),_vm._v(" "),_c('div',{staticClass:"or-rg-bar-mrk",style:({ left: _vm.percentageValue(index) + '%'})})]),_vm._v(" "),(attribute.minLabel || attribute.maxLabel)?_c('div',{staticClass:"or-rg-lbl"},[_c('div',{staticClass:"or-rg-lbl-itm",attrs:{"aria-hidden":"true"}},[_vm._v("\n                        "+_vm._s(attribute.minLabel)+"\n                    ")]),_vm._v(" "),_c('div',{staticClass:"or-rg-lbl-itm",attrs:{"aria-hidden":"true"}},[_vm._v("\n                        "+_vm._s(attribute.maxLabel)+"\n                    ")])]):_vm._e()]):_vm._e(),_vm._v(" "),(attribute.type === 'centered-range')?_c('div',{staticClass:"or-crg"},[_c('span',{staticClass:"okeReviews-a11yText"},[_vm._v("Rated "+_vm._s(_vm.scaledValue(index))+" on a scale of minus 2 to 2, where minus 2 is "+_vm._s(attribute.minLabel)+", 0 is "+_vm._s(attribute.midLabel)+" and 2 is "+_vm._s(attribute.maxLabel))]),_vm._v(" "),_c('div',{staticClass:"or-crg-bar"},[_c('span',{staticClass:"or-crg-bar-bg",style:(_vm.percentageValue(index) < 50 
                            ? { left: _vm.percentageValue(index) + '%', right: '50%' }
                            : { left: '50%', right: 100 - _vm.percentageValue(index) + '%' })}),_vm._v(" "),_c('div',{staticClass:"or-crg-bar-seg"}),_vm._v(" "),_c('div',{staticClass:"or-crg-bar-seg"}),_vm._v(" "),_c('div',{staticClass:"or-crg-bar-mid"}),_vm._v(" "),_c('div',{staticClass:"or-crg-bar-mrk",style:({ left: _vm.percentageValue(index) + '%' })})]),_vm._v(" "),_c('div',{staticClass:"or-crg-lbl"},[_c('div',{staticClass:"or-crg-lbl-itm",attrs:{"aria-hidden":"true"}},[_vm._v("\n                        "+_vm._s(attribute.minLabel)+"\n                    ")]),_vm._v(" "),_c('div',{staticClass:"or-crg-lbl-itm",attrs:{"aria-hidden":"true"}},[_vm._v("\n                        "+_vm._s(attribute.midLabel)+"\n                    ")]),_vm._v(" "),_c('div',{staticClass:"or-crg-lbl-itm",attrs:{"aria-hidden":"true"}},[_vm._v("\n                        "+_vm._s(attribute.maxLabel)+"\n                    ")])])]):_vm._e()])])}),0)};
var __vue_staticRenderFns__$5 = [];

  /* style */
  const __vue_inject_styles__$5 = undefined;
  /* scoped */
  const __vue_scope_id__$5 = undefined;
  /* module identifier */
  const __vue_module_identifier__$5 = undefined;
  /* functional template */
  const __vue_is_functional_template__$5 = false;
  /* style inject */
  
  /* style inject SSR */
  
  /* style inject shadow dom */
  

  
  var OkendoBarAttributes = normalizeComponent(
    { render: __vue_render__$5, staticRenderFns: __vue_staticRenderFns__$5 },
    __vue_inject_styles__$5,
    __vue_script__$5,
    __vue_scope_id__$5,
    __vue_is_functional_template__$5,
    __vue_module_identifier__$5,
    false,
    undefined,
    undefined,
    undefined
  );

//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
var script$6 = {
  props: {
    media: {
      type: Array,
      required: true
    },
    isOverflow: {
      type: Boolean,
      required: false,
      default: false
    }
  },
  methods: {
    onMediaClick(event, clickedIndex) {
      event.preventDefault();
      this.$emit('mediaClick', {
        mediaItems: this.media,
        reviewId: this.media[clickedIndex].reviewId || undefined,
        //aggregate media has reviewId, review media does not
        initialIndex: clickedIndex
      });
    },

    findMediaUrl(mediaItem, urlType = 'thumbnailUrl') {
      if (mediaItem.type === 'image') {
        return mediaItem[urlType] || mediaItem.imageUrls[urlType];
      } else {
        return mediaItem[urlType] || mediaItem.videoUrls[urlType];
      }
    }

  }
};

/* script */
const __vue_script__$6 = script$6;

/* template */
var __vue_render__$6 = function () {var _vm=this;var _h=_vm.$createElement;var _c=_vm._self._c||_h;return _c('div',{staticClass:"okeReviews-mediaStrip",class:{ 'is-overflow': _vm.isOverflow }},[_vm._l((_vm.media),function(mediaItem,index){return [(mediaItem.type === 'image')?_c('a',{key:mediaItem.streamId,staticClass:"okeReviews-mediaStrip-item",attrs:{"href":_vm.findMediaUrl(mediaItem, 'largeUrl'),"role":"button"},on:{"click":function($event){return _vm.onMediaClick($event, index)}}},[_c('img',{staticClass:"okeReviews-mediaStrip-item-media",attrs:{"src":_vm.findMediaUrl(mediaItem, 'thumbnailUrl'),"alt":""}})]):_c('a',{key:mediaItem.streamId,staticClass:"okeReviews-mediaStrip-item",attrs:{"href":_vm.findMediaUrl(mediaItem, 'fullSizeUrl'),"role":"button"},on:{"click":function($event){return _vm.onMediaClick($event, index)}}},[_c('div',{staticClass:"okeReviews-mediaStrip-item-media okeReviews-mediaStrip-item-media--video",style:({ backgroundImage: ("url(" + (_vm.findMediaUrl(mediaItem, 'videoPosterUrl')) + ")") }),attrs:{"aria-label":""}})])]})],2)};
var __vue_staticRenderFns__$6 = [];

  /* style */
  const __vue_inject_styles__$6 = function (inject) {
    if (!inject) return
    inject("data-v-ffd0bc32_0", { source: ".okeReviews .okeReviews-mediaStrip-item{margin-right:.25em}", map: undefined, media: undefined });

  };
  /* scoped */
  const __vue_scope_id__$6 = undefined;
  /* module identifier */
  const __vue_module_identifier__$6 = undefined;
  /* functional template */
  const __vue_is_functional_template__$6 = false;
  /* style inject SSR */
  
  /* style inject shadow dom */
  

  
  var OkendoMediaStrip = normalizeComponent(
    { render: __vue_render__$6, staticRenderFns: __vue_staticRenderFns__$6 },
    __vue_inject_styles__$6,
    __vue_script__$6,
    __vue_scope_id__$6,
    __vue_is_functional_template__$6,
    __vue_module_identifier__$6,
    false,
    createInjector,
    undefined,
    undefined
  );

//
var script$7 = {
  props: {
    itemId: {
      type: String,
      required: true
    },
    parentId: {
      type: String,
      required: false
    },
    helpfulCount: {
      type: Number,
      required: false,
      default: 0
    },
    unhelpfulCount: {
      type: Number,
      required: false,
      default: 0
    },
    contentType: {
      type: 'review' | 'question' | 'answer',
      required: false,
      default: 'review'
    }
  },

  data() {
    return {
      currentVote: undefined,
      isVoting: false
    };
  },

  computed: {
    helpfulCountDisplay() {
      return this.currentVote === 'upvote' ? this.helpfulCount + 1 : this.helpfulCount;
    },

    unhelpfulCountDisplay() {
      return this.currentVote === 'downvote' ? this.unhelpfulCount + 1 : this.unhelpfulCount;
    },

    helpfulLabelText() {
      return this.contentType === 'question' ? 'I had this question too' : 'Was this helpful?';
    }

  },
  methods: {
    onVote(newVote) {
      this.isVoting = true;
      const voteType = this.getVoteType(this.currentVote, newVote);
      const votePromise = this.getVotePromise(voteType);
      votePromise.then(() => {
        this.currentVote = newVote === this.currentVote ? '' : newVote;
        this.isVoting = false;
      }).catch(() => {
        this.isVoting = false;
      });
    },

    getVoteType(currentVote, newVote) {
      if (!currentVote) {
        return newVote;
      } else if (currentVote === newVote) {
        return `${newVote}-cancel`;
      } else {
        return `swap-to-${newVote}`;
      }
    },

    getVotePromise(voteType) {
      switch (this.contentType) {
        case 'review':
          return OkendoAPI.postHelpfulVoteReview(this.itemId, voteType);

        case 'question':
          return OkendoAPI.postHelpfulVoteQuestion(this.itemId, voteType);

        case 'answer':
          return OkendoAPI.postHelpfulVoteAnswer(this.parentId, this.itemId, voteType);
      }
    }

  }
};

/* script */
const __vue_script__$7 = script$7;

/* template */
var __vue_render__$7 = function () {var _vm=this;var _h=_vm.$createElement;var _c=_vm._self._c||_h;return _c('div',{staticClass:"oke-helpful"},[_c('span',{staticClass:"oke-helpful-label"},[_vm._v(_vm._s(_vm.helpfulLabelText))]),_vm._v(" "),_c('div',{staticClass:"oke-helpful-vote"},[_c('button',{staticClass:"oke-helpful-vote-button oke-helpful-vote-button--positive",class:{ 'oke-is-voted': _vm.currentVote === 'upvote', 'is-oke-disabled': _vm.isVoting },attrs:{"disabled":_vm.isVoting,"aria-describedby":("oke-helpful-vote-counter-positive-" + _vm.itemId),"type":"button","aria-label":"Yes"},on:{"click":function($event){return _vm.onVote('upvote')}}},[_c('span',{staticClass:"oke-helpful-vote-counter",attrs:{"id":("oke-helpful-vote-counter-positive-" + _vm.itemId)}},[_vm._v("\n                "+_vm._s(_vm.helpfulCountDisplay)+" "),_c('span',{staticClass:"okeReviews-a11yText"},[_vm._v(_vm._s(_vm.helpfulCountDisplay === 1 ? 'person voted yes' : 'people voted yes'))])])]),_vm._v(" "),_c('button',{staticClass:"oke-helpful-vote-button oke-helpful-vote-button--negative",class:{ 'oke-is-voted': _vm.currentVote === 'downvote', 'is-oke-disabled': _vm.isVoting },attrs:{"disabled":_vm.isVoting,"aria-describedby":("oke-helpful-vote-counter-negative-" + _vm.itemId),"type":"button","aria-label":"No"},on:{"click":function($event){return _vm.onVote('downvote')}}},[_c('span',{staticClass:"oke-helpful-vote-counter",attrs:{"id":("oke-helpful-vote-counter-negative-" + _vm.itemId)}},[_vm._v("\n                "+_vm._s(_vm.unhelpfulCountDisplay)+" "),_c('span',{staticClass:"okeReviews-a11yText"},[_vm._v(_vm._s(_vm.unhelpfulCountDisplay === 1 ? 'person voted no' : 'people voted no'))])])])])])};
var __vue_staticRenderFns__$7 = [];

  /* style */
  const __vue_inject_styles__$7 = function (inject) {
    if (!inject) return
    inject("data-v-96f199c4_0", { source: ".oke-helpful{color:#7c7c7c;display:block;font-size:12px}.oke-helpful-question{vertical-align:middle}.oke-helpful-vote{display:inline-block;margin-left:1em}.oke-helpful-vote-button{background:0 0;border:none;color:#7c7c7c;cursor:pointer;padding:0;transition:opacity .3s ease;vertical-align:middle;width:auto}.oke-helpful-vote-button::before{vertical-align:middle;margin-bottom:.3em}.oke-helpful-vote-button--positive::before{background-image:url(\"data:image/svg+xml,%3Csvg fill='%237c7c7c' xmlns='http://www.w3.org/2000/svg' viewBox='0 0 24 24'%3E%3Cpath d='M23,10C23,8.89 22.1,8 21,8H14.68L15.64,3.43C15.66,3.33 15.67,3.22 15.67,3.11C15.67,2.7 15.5,2.32 15.23,2.05L14.17,1L7.59,7.58C7.22,7.95 7,8.45 7,9V19A2,2 0 0,0 9,21H18C18.83,21 19.54,20.5 19.84,19.78L22.86,12.73C22.95,12.5 23,12.26 23,12V10M1,21H5V9H1V21Z' /%3E%3C/svg%3E\");background-repeat:no-repeat;display:inline-block;content:\"\";width:1em;height:1em;background-size:1em 1em}.oke-helpful-vote-button--negative::before{background-image:url(\"data:image/svg+xml,%3Csvg fill='%237c7c7c' xmlns='http://www.w3.org/2000/svg' viewBox='0 0 24 24'%3E%3Cpath d='M19,15H23V3H19M15,3H6C5.17,3 4.46,3.5 4.16,4.22L1.14,11.27C1.05,11.5 1,11.74 1,12V14A2,2 0 0,0 3,16H9.31L8.36,20.57C8.34,20.67 8.33,20.77 8.33,20.88C8.33,21.3 8.5,21.67 8.77,21.94L9.83,23L16.41,16.41C16.78,16.05 17,15.55 17,15V5C17,3.89 16.1,3 15,3Z' /%3E%3C/svg%3E\");background-repeat:no-repeat;display:inline-block;content:\"\";width:1em;height:1em;background-size:1em 1em;margin-left:1em}.oke-helpful-vote-button--negative::before,.oke-helpful-vote-button--positive::before{opacity:.5}.oke-helpful-vote-button--negative.oke-is-voted::before,.oke-helpful-vote-button--positive.oke-is-voted::before{color:inherit;opacity:1}.oke-helpful-vote-button.is-oke-disabled{opacity:.3}.oke-helpful-vote-counter{display:inline-block;margin-left:.25em;pointer-events:none}", map: undefined, media: undefined });

  };
  /* scoped */
  const __vue_scope_id__$7 = undefined;
  /* module identifier */
  const __vue_module_identifier__$7 = undefined;
  /* functional template */
  const __vue_is_functional_template__$7 = false;
  /* style inject SSR */
  
  /* style inject shadow dom */
  

  
  var HelpfulVoting = normalizeComponent(
    { render: __vue_render__$7, staticRenderFns: __vue_staticRenderFns__$7 },
    __vue_inject_styles__$7,
    __vue_script__$7,
    __vue_scope_id__$7,
    __vue_is_functional_template__$7,
    __vue_module_identifier__$7,
    false,
    createInjector,
    undefined,
    undefined
  );

//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
var script$8 = {
  props: {
    productName: {
      type: String,
      required: true
    },
    productImageUrl: {
      type: String,
      required: false
    },
    productUrl: {
      type: String,
      required: false
    }
  }
};

/* script */
const __vue_script__$8 = script$8;

/* template */
var __vue_render__$8 = function () {var _vm=this;var _h=_vm.$createElement;var _c=_vm._self._c||_h;return _c('div',{staticClass:"okeReviews-review-reviewer-product"},[(_vm.productImageUrl && _vm.productUrl)?_c('a',{staticClass:"okeReviews-review-reviewer-product-thumbnail",attrs:{"href":_vm.productUrl}},[_c('img',{staticClass:"okeReviews-review-reviewer-product-thumbnail-image",attrs:{"src":_vm.productImageUrl,"alt":_vm.productName}})]):_c('div',{staticClass:"okeReviews-review-reviewer-product-thumbnail"},[_c('img',{staticClass:"okeReviews-review-reviewer-product-thumbnail-image",attrs:{"src":_vm.productImageUrl,"alt":_vm.productName}})]),_vm._v(" "),_c('div',{staticClass:"okeReviews-review-reviewer-product-details"},[_c('div',{staticClass:"okeReviews-review-reviewer-product-details-heading"},[_vm._v("\n            Reviewing\n        ")]),_vm._v(" "),(_vm.productUrl)?_c('a',{staticClass:"okeReviews-review-reviewer-product-details-link",attrs:{"href":_vm.productUrl}},[_vm._v("\n            "+_vm._s(_vm.productName)+"\n        ")]):_c('div',[_vm._v("\n            "+_vm._s(_vm.productName)+"\n        ")])])])};
var __vue_staticRenderFns__$8 = [];

  /* style */
  const __vue_inject_styles__$8 = undefined;
  /* scoped */
  const __vue_scope_id__$8 = undefined;
  /* module identifier */
  const __vue_module_identifier__$8 = undefined;
  /* functional template */
  const __vue_is_functional_template__$8 = false;
  /* style inject */
  
  /* style inject SSR */
  
  /* style inject shadow dom */
  

  
  var OkendoProduct = normalizeComponent(
    { render: __vue_render__$8, staticRenderFns: __vue_staticRenderFns__$8 },
    __vue_inject_styles__$8,
    __vue_script__$8,
    __vue_scope_id__$8,
    __vue_is_functional_template__$8,
    __vue_module_identifier__$8,
    false,
    undefined,
    undefined,
    undefined
  );

function createHandler(divisor,noun,restOfString){
	return function(diff){
		var n = Math.floor(diff/divisor);
		var pluralizedNoun = noun + ( n > 1 ? 's' : '' );
		return "" + n + " " + pluralizedNoun + " " + restOfString;
	}
}

var formatters = [
	{ threshold: 1,        handler: function(){ return      "just now" } },
	{ threshold: 60,       handler: createHandler(1,        "second",    "ago" ) },
	{ threshold: 3600,     handler: createHandler(60,       "minute",    "ago" ) },
	{ threshold: 86400,    handler: createHandler(3600,     "hour",      "ago" ) },
	{ threshold: 172800,   handler: function(){ return      "yesterday" } },
	{ threshold: 604800,   handler: createHandler(86400,    "day",       "ago" ) },
	{ threshold: 2592000,  handler: createHandler(604800,   "week",      "ago" ) },
	{ threshold: 31536000, handler: createHandler(2592000,  "month",     "ago" ) },
	{ threshold: Infinity, handler: createHandler(31536000, "year",      "ago" ) }
];

var prettydate = {
	format: function (date) {
		var diff = (((new Date()).getTime() - date.getTime()) / 1000);
		for( var i=0; i<formatters.length; i++ ){
			if( diff < formatters[i].threshold ){
				return formatters[i].handler(diff);
			}
		}
		throw new Error("exhausted all formatter options, none found"); //should never be reached
	}
};
 
var prettyDate = prettydate;

var OkendoUtils = {
  getPrettyDate(date) {
    return prettyDate.format(new Date(date));
  },

  addParagraphsTags(text) {
    let encodedParagraphs = '<p>';
    encodedParagraphs += text.replace(/(?:\r\n|\r|\n)+/g, '</p><p>');
    encodedParagraphs += '</p>';
    return encodedParagraphs;
  }

};

//
var script$9 = {
  props: {
    review: {
      type: Object,
      required: true
    },
    showProductData: {
      type: Boolean,
      required: false,
      default: false
    }
  },
  components: {
    OkendoCustomerAttributes,
    OkendoProduct,
    OkendoProductAttributes,
    OkendoReviewerProfile,
    OkendoStars,
    OkendoBarAttributes,
    OkendoMediaStrip,
    HelpfulVoting
  },

  data() {
    return {
      formattedDate: '',
      reviewBodyParagraphs: ''
    };
  },

  mounted() {
    this.formattedDate = OkendoUtils.getPrettyDate(this.review.dateCreated);
    this.reviewBodyParagraphs = OkendoUtils.addParagraphsTags(this.review.body);
  },

  methods: {
    openMediaModal(mediaModalMessage) {
      this.$emit('mediaClick', { ...mediaModalMessage,
        reviewId: this.review.reviewId,
        source: 'review'
      });
    }

  }
};

/* script */
const __vue_script__$9 = script$9;

/* template */
var __vue_render__$9 = function () {var _vm=this;var _h=_vm.$createElement;var _c=_vm._self._c||_h;return _c('article',{staticClass:"okeReviews-review"},[_c('aside',{staticClass:"okeReviews-review-side"},[_c('div',{staticClass:"okeReviews-review-side-inner"},[_c('div',{staticClass:"okeReviews-review-reviewer"},[_c('okendo-reviewer-profile',{attrs:{"reviewer":_vm.review.reviewer}}),_vm._v(" "),(_vm.showProductData)?_c('okendo-product',{attrs:{"productName":_vm.review.productName,"productUrl":_vm.review.productUrl,"productImageUrl":_vm.review.productImageUrl}}):_vm._e(),_vm._v(" "),_c('div',{staticClass:"okeReviews-review-recommendation",class:_vm.review.isRecommended ? 'okeReviews-review-recommendation--yes' : 'okeReviews-review-recommendation--no'},[_vm._v("\n                    "+_vm._s(_vm.review.isRecommended ? 'I recommend this product' : 'I don\'t recommend this product')+"\n                ")]),_vm._v(" "),(_vm.review.reviewer.attributes)?_c('okendo-customer-attributes',{attrs:{"attributes":_vm.review.reviewer.attributes}}):_vm._e()],1),_vm._v(" "),(_vm.review.productAttributes)?_c('okendo-product-attributes',{attrs:{"attributes":_vm.review.productAttributes}}):_vm._e()],1)]),_vm._v(" "),_c('div',{staticClass:"okeReviews-review-primary"},[_c('div',{staticClass:"okeReviews-review-meta"},[_c('div',{staticClass:"okeReviews-review-starRating"},[_c('okendo-stars',{attrs:{"rating":_vm.review.rating}})],1),_vm._v(" "),_c('div',{staticClass:"okeReviews-review-date"},[_c('span',{staticClass:"okeReviews-a11yText"},[_vm._v("Review posted ")]),_c('span',[_vm._v(_vm._s(_vm.formattedDate))])])]),_vm._v(" "),_c('div',{staticClass:"okeReviews-review-main"},[_c('h2',{staticClass:"okeReviews-review-main-heading"},[_vm._v(_vm._s(_vm.review.title))]),_vm._v(" "),_c('div',{staticClass:"okeReviews-review-main-content"},[_c('div',{staticClass:"okeReviews-review-main-content-body",domProps:{"innerHTML":_vm._s(_vm.reviewBodyParagraphs)}})])]),_vm._v(" "),(_vm.review.attributesWithRating)?_c('div',{staticClass:"okeReviews-review-attributeRatings"},[_c('table',{staticClass:"okeReviews-attributeRatingsTable",attrs:{"aria-label":"Product attributes ratings"}},[_c('okendo-bar-attributes',{attrs:{"attributes":_vm.review.attributesWithRating}})],1)]):_vm._e(),_vm._v(" "),(_vm.review.media && _vm.review.media.length)?_c('div',{staticClass:"okeReviews-review-media"},[_c('okendo-media-strip',{attrs:{"media":_vm.review.media},on:{"mediaClick":function($event){return _vm.openMediaModal($event)}}})],1):_vm._e(),_vm._v(" "),_c('helpful-voting',{attrs:{"itemId":_vm.review.reviewId,"helpfulCount":_vm.review.helpfulCount,"unhelpfulCount":_vm.review.unhelpfulCount,"contentType":"review"}})],1)])};
var __vue_staticRenderFns__$9 = [];

  /* style */
  const __vue_inject_styles__$9 = undefined;
  /* scoped */
  const __vue_scope_id__$9 = undefined;
  /* module identifier */
  const __vue_module_identifier__$9 = undefined;
  /* functional template */
  const __vue_is_functional_template__$9 = false;
  /* style inject */
  
  /* style inject SSR */
  
  /* style inject shadow dom */
  

  
  var OkendoReview = normalizeComponent(
    { render: __vue_render__$9, staticRenderFns: __vue_staticRenderFns__$9 },
    __vue_inject_styles__$9,
    __vue_script__$9,
    __vue_scope_id__$9,
    __vue_is_functional_template__$9,
    __vue_module_identifier__$9,
    false,
    undefined,
    undefined,
    undefined
  );

//
//
//
//
//
//
//
//
//
//
//
//
var script$a = {
  props: {
    url: {
      type: String,
      required: true
    }
  }
};

/* script */
const __vue_script__$a = script$a;

/* template */
var __vue_render__$a = function () {var _vm=this;var _h=_vm.$createElement;var _c=_vm._self._c||_h;return (_vm.url)?_c('a',{staticClass:"okeReviews-writeReviewButton okeReviews-button",attrs:{"href":_vm.url,"target":"_blank","rel":"nofollow"}},[_vm._v("\n    Write a Review\n")]):_vm._e()};
var __vue_staticRenderFns__$a = [];

  /* style */
  const __vue_inject_styles__$a = function (inject) {
    if (!inject) return
    inject("data-v-3f97dd42_0", { source: ".okeReviews .okeReviews-writeReviewButton{visibility:visible;border:0;padding-left:24px;padding-right:24px;background-color:#f5f5f5;display:inline-block}", map: undefined, media: undefined });

  };
  /* scoped */
  const __vue_scope_id__$a = undefined;
  /* module identifier */
  const __vue_module_identifier__$a = undefined;
  /* functional template */
  const __vue_is_functional_template__$a = false;
  /* style inject SSR */
  
  /* style inject shadow dom */
  

  
  var OkendoWriteReviewButton = normalizeComponent(
    { render: __vue_render__$a, staticRenderFns: __vue_staticRenderFns__$a },
    __vue_inject_styles__$a,
    __vue_script__$a,
    __vue_scope_id__$a,
    __vue_is_functional_template__$a,
    __vue_module_identifier__$a,
    false,
    createInjector,
    undefined,
    undefined
  );

//
var script$b = {
  props: {
    viewState: {
      type: String,
      required: true
    },
    writeReviewUrl: {
      type: String
    }
  },
  components: {
    OkendoWriteReviewButton
  }
};

/* script */
const __vue_script__$b = script$b;

/* template */
var __vue_render__$b = function () {var _vm=this;var _h=_vm.$createElement;var _c=_vm._self._c||_h;return _c('div',{staticClass:"okeReviews-reviewsWidget-stateMessage"},[(_vm.viewState === 'error')?[_vm._v("Whoops, something went wrong 😣")]:_vm._e(),_vm._v(" "),(_vm.viewState === 'loading')?[_vm._v("Loading Reviews...")]:_vm._e(),_vm._v(" "),(_vm.viewState === 'empty')?[_vm._v("\n        No reviews yet"),(_vm.writeReviewUrl)?[_vm._v(", write one now?")]:_vm._e(),_vm._v(" "),(_vm.writeReviewUrl)?_c('div',{staticClass:"okeReviews-reviewsWidget-stateMessage-button"},[_c('okendo-write-review-button',{attrs:{"url":_vm.writeReviewUrl}})],1):_vm._e()]:_vm._e()],2)};
var __vue_staticRenderFns__$b = [];

  /* style */
  const __vue_inject_styles__$b = function (inject) {
    if (!inject) return
    inject("data-v-6b82f7dd_0", { source: ".okeReviews-reviewsWidget-stateMessage{font-family:inherit;border-top:2px solid #f5f5f5;border-bottom:2px solid #f5f5f5;padding:24px;text-align:center}.okeReviews-reviewsWidget-stateMessage-button{margin-top:8px}", map: undefined, media: undefined });

  };
  /* scoped */
  const __vue_scope_id__$b = undefined;
  /* module identifier */
  const __vue_module_identifier__$b = undefined;
  /* functional template */
  const __vue_is_functional_template__$b = false;
  /* style inject SSR */
  
  /* style inject shadow dom */
  

  
  var OkendoState = normalizeComponent(
    { render: __vue_render__$b, staticRenderFns: __vue_staticRenderFns__$b },
    __vue_inject_styles__$b,
    __vue_script__$b,
    __vue_scope_id__$b,
    __vue_is_functional_template__$b,
    __vue_module_identifier__$b,
    false,
    createInjector,
    undefined,
    undefined
  );

//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
var script$c = {
  props: {
    reviewCountByLevel: {
      type: Object,
      required: true
    },
    reviewCountTotal: {
      type: Number,
      requred: true
    }
  },

  data() {
    return {
      reviewCountData: undefined
    };
  },

  mounted() {
    this.reviewCountData = {
      level5Count: this.reviewCountByLevel.level5Count || 0,
      level4Count: this.reviewCountByLevel.level4Count || 0,
      level3Count: this.reviewCountByLevel.level3Count || 0,
      level2Count: this.reviewCountByLevel.level2Count || 0,
      level1Count: this.reviewCountByLevel.level1Count || 0
    };
  },

  methods: {
    reviewCountPercentage(count) {
      return count / this.reviewCountTotal * 100 + '%';
    },

    transformIndex(index) {
      return 5 - index;
    }

  }
};

/* script */
const __vue_script__$c = script$c;

/* template */
var __vue_render__$c = function () {var _vm=this;var _h=_vm.$createElement;var _c=_vm._self._c||_h;return _c('div',{staticClass:"okeReviews-reviewsAggregate-ratingDistribution"},[_c('ul',{staticClass:"okeReviews-ratingDistribution",attrs:{"aria-label":"Rating distribution"}},_vm._l((_vm.reviewCountData),function(count,key,index){return _c('li',{key:key,staticClass:"okeReviews-ratingDistribution-item"},[_c('div',{staticClass:"okeReviews-ratingDistribution-item-label"},[_c('div',{staticClass:"okeReviews-ratingDistribution-item-label-inner"},[_vm._v(_vm._s(_vm.transformIndex(index))+" "+_vm._s(count === 1 ? "Star" : "Stars"))])]),_vm._v(" "),_c('div',{staticClass:"okeReviews-ratingDistribution-item-value"},[_c('div',{staticClass:"okeReviews-ratingDistribution-item-value-inner"},[_c('div',{staticClass:"okeReviews-ratingDistribution-item-value-graph",attrs:{"role":"presentation"}},[_c('div',{staticClass:"okeReviews-barGraph"},[_c('div',{staticClass:"okeReviews-barGraph-layer okeReviews-barGraph-layer--background"}),_vm._v(" "),_c('div',{staticClass:"okeReviews-barGraph-layer okeReviews-barGraph-layer--foreground",style:({ width: _vm.reviewCountPercentage(count) })})])]),_vm._v(" "),_c('div',{staticClass:"okeReviews-ratingDistribution-item-value-count"},[_vm._v("\n                        "+_vm._s(count)+" "),_c('span',{staticClass:"okeReviews-a11yText"},[_vm._v(_vm._s(count === 1 ? "Review" : "Reviews"))])])])])])}),0)])};
var __vue_staticRenderFns__$c = [];

  /* style */
  const __vue_inject_styles__$c = undefined;
  /* scoped */
  const __vue_scope_id__$c = undefined;
  /* module identifier */
  const __vue_module_identifier__$c = undefined;
  /* functional template */
  const __vue_is_functional_template__$c = false;
  /* style inject */
  
  /* style inject SSR */
  
  /* style inject shadow dom */
  

  
  var OkendoRatingBreakdown = normalizeComponent(
    { render: __vue_render__$c, staticRenderFns: __vue_staticRenderFns__$c },
    __vue_inject_styles__$c,
    __vue_script__$c,
    __vue_scope_id__$c,
    __vue_is_functional_template__$c,
    __vue_module_identifier__$c,
    false,
    undefined,
    undefined,
    undefined
  );

//
var script$d = {
  props: {
    aggregate: {
      type: Object,
      required: true
    },
    mediaAggregate: {
      type: Object
    }
  },
  components: {
    OkendoStars,
    OkendoRatingBreakdown,
    OkendoBarAttributes,
    OkendoMediaStrip
  },
  computed: {
    percentageRecommended() {
      return ((this.aggregate.recommendationCount || 0) / this.aggregate.reviewCount * 100).toFixed(0);
    },

    averageRating() {
      return (this.aggregate.ratingAndReviewValuesTotal / this.aggregate.ratingAndReviewCount).toFixed(1);
    },

    ratingsEnabled() {
      return !!this.aggregate.ratingCount;
    },

    hasMedia() {
      return this.mediaAggregate && this.mediaAggregate.media && this.mediaAggregate.media.length;
    }

  },
  methods: {
    openMediaModal(mediaModalMessage) {
      this.$emit('mediaClick', { ...mediaModalMessage,
        source: 'aggregate'
      });
    }

  }
};

/* script */
const __vue_script__$d = script$d;

/* template */
var __vue_render__$d = function () {var _vm=this;var _h=_vm.$createElement;var _c=_vm._self._c||_h;return _vm._m(0)};
var __vue_staticRenderFns__$d = [function () {var _vm=this;var _h=_vm.$createElement;var _c=_vm._self._c||_h;return _c('div',{staticClass:"okeReviews-reviewsWidget-reviewsAggregate"},[_c('div',{staticClass:"okeReviews-reviewsAggregate"},[_c('div',{staticClass:"okeReviews-reviewsAggregate-side"},[_c('div',{staticClass:"okeReviews-reviewsAggregate-side-inner"},[_c('div',{staticClass:"okeReviews-reviewsAggregate-summary"},[_c('div',{staticClass:"okeReviews-reviewsAggregate-summary-rating"},[_c('div',{staticClass:"okeReviews-reviewsAggregate-summary-rating-value"},[_c('div',{staticClass:"okeReviews-badge okeReviews-badge--large okeReviews-badge--rating"},[_c('span',{staticClass:"okeReviews-a11yText"},[_vm._v("Average Rating")]),_vm._v(" "),_c('span',[_vm._v(_vm._s(_vm.averageRating))]),_vm._v(" "),_c('span',{staticClass:"okeReviews-a11yText"},[_vm._v("out of 5")])])]),_vm._v(" "),_c('div',{staticClass:"okeReviews-reviewsAggregate-summary-rating-starRating",attrs:{"role":"presentation"}},[_c('okendo-stars',{attrs:{"rating":_vm.averageRating}})],1)]),_vm._v(" "),_c('div',{staticClass:"okeReviews-reviewsAggregate-summary-total"},[(_vm.ratingsEnabled)?[_vm._v("\n                            Based on "+_vm._s(_vm.aggregate.ratingAndReviewCount || '0')+" "+_vm._s(_vm.aggregate.ratingAndReviewCount === 1 ? 'Rating' : 'Ratings')+"\n                        ")]:[_vm._v("\n                            Based on "+_vm._s(_vm.aggregate.ratingAndReviewCount || '0')+" "+_vm._s(_vm.aggregate.ratingAndReviewCount === 1 ? 'Review' : 'Reviews')+"\n                        ")]],2)]),_vm._v(" "),(_vm.aggregate.attributeRatings || _vm.hasMedia)?_c('okendo-rating-breakdown',{attrs:{"reviewCountTotal":_vm.aggregate.ratingAndReviewCount,"reviewCountByLevel":_vm.aggregate.ratingAndReviewCountByLevel}}):_vm._e()],1)]),_vm._v(" "),_c('div',{staticClass:"okeReviews-reviewsAggregate-primary"},[(_vm.percentageRecommended)?_c('div',{staticClass:"okeReviews-reviewsAggregate-recommends"},[_c('span',{staticClass:"okeReviews-reviewsAggregate-recommends-badge okeReviews-badge okeReviews-badge--recommended"},[_vm._v(_vm._s(_vm.percentageRecommended)+"%")]),_vm._v(" "),_c('span',[_vm._v(_vm._s(_vm.aggregate.productName ? 'of reviewers would recommend this product to a friend' :  'of reviewers would recommend these products to a friend'))])]):_vm._e(),_vm._v(" "),_c('div',{staticClass:"okeReviews-reviewsAggregate-mod"},[(_vm.aggregate.attributeRatings)?_c('div',{staticClass:"okeReviews-reviewsAggregate-mod-itm okeReviews-reviewsAggregate-mod-itm--attRtgAv"},[_c('table',{staticClass:"okeReviews-attributeRatingsTable",attrs:{"aria-label":"Product attribute rating averages"}},[_c('okendo-bar-attributes',{attrs:{"attributes":_vm.aggregate.attributeRatings}})],1)]):_vm._e(),_vm._v(" "),(_vm.hasMedia)?_c('div',{staticClass:"okeReviews-reviewsAggregate-mod-itm okeReviews-reviewsAggregate-mod-itm--medAgg"},[_c('div',{staticClass:"okeReviews-mediaAggregate"},[_c('span',{staticClass:"okeReviews-a11yText"},[_vm._v("Customer photos and videos")]),_vm._v(" "),_c('okendo-media-strip',{attrs:{"media":_vm.mediaAggregate.media,"isOverflow":!!_vm.mediaAggregate.nextUrl},on:{"mediaClick":function($event){return _vm.openMediaModal($event)}}})],1)]):_vm._e()])])])])}];

  /* style */
  const __vue_inject_styles__$d = undefined;
  /* scoped */
  const __vue_scope_id__$d = undefined;
  /* module identifier */
  const __vue_module_identifier__$d = undefined;
  /* functional template */
  const __vue_is_functional_template__$d = false;
  /* style inject */
  
  /* style inject SSR */
  
  /* style inject shadow dom */
  

  
  var OkendoAggregate = normalizeComponent(
    { render: __vue_render__$d, staticRenderFns: __vue_staticRenderFns__$d },
    __vue_inject_styles__$d,
    __vue_script__$d,
    __vue_scope_id__$d,
    __vue_is_functional_template__$d,
    __vue_module_identifier__$d,
    false,
    undefined,
    undefined,
    undefined
  );

//
var script$e = {
  props: {
    review: {
      type: Object,
      required: true
    },
    showProductData: {
      type: Boolean,
      required: false,
      default: false
    }
  },
  components: {
    OkendoReviewerProfile,
    OkendoStars,
    HelpfulVoting,
    OkendoProduct
  },

  data() {
    return {
      formattedDate: '',
      reviewBodyParagraphs: ''
    };
  },

  mounted() {
    this.formattedDate = OkendoUtils.getPrettyDate(this.review.dateCreated);
    this.reviewBodyParagraphs = OkendoUtils.addParagraphsTags(this.review.body);
  }

};

/* script */
const __vue_script__$e = script$e;

/* template */
var __vue_render__$e = function () {var _vm=this;var _h=_vm.$createElement;var _c=_vm._self._c||_h;return _c('div',{staticClass:"okeReviews-review okeReviews-review--simple"},[_c('okendo-reviewer-profile',{attrs:{"reviewer":_vm.review.reviewer}}),_vm._v(" "),(_vm.showProductData)?_c('okendo-product',{attrs:{"productName":_vm.review.productName,"productUrl":_vm.review.productUrl,"productImageUrl":_vm.review.productImageUrl}}):_vm._e(),_vm._v(" "),_c('div',{staticClass:"okeReviews-review-meta"},[_c('div',{staticClass:"okeReviews-review-date"},[_c('span',{staticClass:"okeReviews-a11yText"},[_vm._v("Review posted ")]),_c('span',[_vm._v(_vm._s(_vm.formattedDate))])]),_vm._v(" "),_c('div',{staticClass:"okeReviews-review-starRating"},[_c('okendo-stars',{attrs:{"rating":_vm.review.rating}})],1)]),_vm._v(" "),_c('div',{staticClass:"okeReviews-review-main"},[_c('h2',{staticClass:"okeReviews-review-main-heading"},[_vm._v(_vm._s(_vm.review.title))]),_vm._v(" "),_c('div',{staticClass:"okeReviews-review-main-content",domProps:{"innerHTML":_vm._s(_vm.reviewBodyParagraphs)}})]),_vm._v(" "),_c('helpful-voting',{attrs:{"itemId":_vm.review.reviewId,"helpfulCount":_vm.review.helpfulCount,"unhelpfulCount":_vm.review.unhelpfulCount,"contentType":"review"}})],1)};
var __vue_staticRenderFns__$e = [];

  /* style */
  const __vue_inject_styles__$e = undefined;
  /* scoped */
  const __vue_scope_id__$e = undefined;
  /* module identifier */
  const __vue_module_identifier__$e = undefined;
  /* functional template */
  const __vue_is_functional_template__$e = false;
  /* style inject */
  
  /* style inject SSR */
  
  /* style inject shadow dom */
  

  
  var OkendoReviewSimple = normalizeComponent(
    { render: __vue_render__$e, staticRenderFns: __vue_staticRenderFns__$e },
    __vue_inject_styles__$e,
    __vue_script__$e,
    __vue_scope_id__$e,
    __vue_is_functional_template__$e,
    __vue_module_identifier__$e,
    false,
    undefined,
    undefined,
    undefined
  );

//
var script$f = {
  components: {
    OkendoReviewSimple
  },
  props: {
    shopifyProductId: {
      type: String,
      required: false
    },
    mediaItems: {
      type: Array,
      required: false
    },
    initialReviewId: {
      type: String,
      required: false
    },
    initialIndex: {
      type: Number,
      default: 0,
      required: false
    },
    mode: {
      type: String,
      required: true
    },
    isAllReviews: {
      type: Boolean,
      required: false,
      default: false
    }
  },

  data() {
    return {
      isMediaLoading: true,
      isLoadingMore: false,
      isError: false,
      media: [],
      nextUrl: '',
      imagesPerPage: 25,
      loadMoreThreshold: 8,
      review: undefined,
      reviewId: ''
    };
  },

  mounted() {
    this.loadMediaGalleryData();
  },

  methods: {
    loadMediaGalleryData() {
      this.reviewId = this.initialReviewId;
      this.loadReview();

      if (this.mode === 'aggregate') {
        const mediaAggregatePromise = this.isAllReviews ? OkendoAPI.getMediaAll(this.imagesPerPage) : OkendoAPI.getMediaProduct(this.shopifyProductId, this.imagesPerPage);
        mediaAggregatePromise.then(mediaResponse => {
          this.media = this.remapMedia(mediaResponse.data.media);
          this.nextUrl = mediaResponse.data.nextUrl;
          setTimeout(() => this.initFlickity(), 0);
          this.isMediaLoading = false;
        }).catch(() => {
          this.isMediaLoading = false;
          this.isError = true;
        });
      } else {
        this.media = this.remapMedia(this.mediaItems);
        this.isMediaLoading = false;
        setTimeout(() => this.initFlickity(), 0);
      }
    },

    loadReview() {
      OkendoAPI.getReview(this.reviewId).then(reviewResponse => {
        this.review = reviewResponse.data.review;
      });
    },

    remapMedia(media, append = false) {
      if (this.mode === 'aggregate') {
        return media.map(medium => ({
          url: medium.type === 'image' ? medium.imageUrls.largeUrl : medium.videoUrls.fullSizeUrl,
          type: medium.type,
          thumbnailUrl: medium.type === 'image' ? medium.imageUrls.thumbnailUrl : medium.videoUrls.thumbnailUrl,
          videoPosterURL: medium.type === 'image' ? undefined : medium.videoUrls.videoPosterUrl,
          streamId: medium.streamId,
          reviewId: medium.reviewId,
          append
        }));
      } else {
        return media.map(medium => ({
          url: medium.type === 'image' ? medium.largeUrl : medium.fullSizeUrl,
          type: medium.type,
          thumbnailUrl: medium.thumbnailUrl,
          videoPosterURL: medium.type === 'image' ? undefined : medium.thumbnailUrl,
          reviewId: this.reviewId,
          append
        }));
      }
    },

    initFlickity() {
      const mediaGallerySliderElement = this.$refs.slider;
      const mediaGallerySliderNavElement = this.$refs.sliderNav;
      const mediaGallerySlider = new Flickity(mediaGallerySliderElement, {
        initialIndex: this.initialIndex,
        accessibility: true,
        prevNextButtons: this.media.length > 1,
        pageDots: false,
        wrapAround: this.mode !== 'aggregate'
      });

      if (this.mode === 'aggregate') {
        const mediaGallerySliderNav = new Flickity(mediaGallerySliderNavElement, {
          asNavFor: mediaGallerySliderElement,
          accessibility: false,
          pageDots: false,
          prevNextButtons: this.media.length > 1
        });
        mediaGallerySlider.on('change', () => {
          const newReviewId = this.media[mediaGallerySliderNav.selectedIndex].reviewId;

          if (newReviewId !== this.reviewId) {
            this.review = undefined;
            this.reviewId = newReviewId;
            this.loadReview();
          }
        });
        mediaGallerySliderNav.on('settle', () => {
          if (this.media && this.nextUrl && mediaGallerySliderNav.cells.length - mediaGallerySliderNav.selectedIndex <= this.loadMoreThreshold && !this.isLoadingMore) {
            this.isLoadingMore = true;
            OkendoAPI.getMore(this.nextUrl).then(response => {
              this.media = this.media.concat(this.remapMedia(response.data.media, true));
              this.nextUrl = response.data.nextUrl;
              setTimeout(() => {
                mediaGallerySliderNav.append(this.$refs.navSlideToAppend);
                mediaGallerySlider.append(this.$refs.mainSlideToAppend);
                this.cleanseMediaOfAppends();
              }, 0);
              this.isLoadingMore = false;
            }).catch(() => this.isLoadingMore = false);
          }
        });
      }

      mediaGallerySliderElement.focus();
    },

    cleanseMediaOfAppends() {
      this.media.map(mediaItem => mediaItem.append = false);
    },

    onClose(event) {
      this.$emit('close');
    }

  }
};

/* script */
const __vue_script__$f = script$f;

/* template */
var __vue_render__$f = function () {var _vm=this;var _h=_vm.$createElement;var _c=_vm._self._c||_h;return _c('div',{staticClass:"okeReviews okeReviews--theme"},[_c('div',{staticClass:"okeReviews-modal"},[_c('div',{staticClass:"okeReviews-modal-overlay",attrs:{"tabindex":"-1"}}),_vm._v(" "),_c('div',{staticClass:"okeReviews-modal-inner",on:{"click":function($event){return _vm.onClose()}}},[_c('div',{staticClass:"okeReviews-modal-dialog",attrs:{"role":"dialog"}},[_c('div',{attrs:{"role":"document"}},[_c('div',{staticClass:"okeReviews-mediaGalleryReviewDialog",on:{"click":function($event){$event.stopPropagation();}}},[(_vm.isMediaLoading)?_c('span',{staticClass:"okeReviews-a11yText",attrs:{"role":"alertdialog","aria-busy":"true"}},[_vm._v("Loading...")]):_vm._e(),_vm._v(" "),_c('div',{staticClass:"okeReviews-mediaGalleryReviewDialog-main"},[_c('div',{staticClass:"okeReviews-mediaGalleryReviewDialog-slider"},[_c('div',{staticClass:"okeReviews-mediaGalleryReviewDialog-slider-inner"},[_c('div',{ref:"slider",staticClass:"okeReviews-mediaSlider",class:{ 'is-loading': _vm.isMediaLoading }},_vm._l((_vm.media),function(mediaItem){return _c('div',{key:mediaItem.thumbnailUrl,ref:mediaItem.append ? 'mainSlideToAppend' : undefined,refInFor:true,staticClass:"okeReviews-mediaSlider-slide"},[(mediaItem.type === 'image')?_c('img',{staticClass:"okeReviews-mediaSlider-slide-media",attrs:{"src":mediaItem.url,"alt":"Customer {{ mediaItem.type }}"}}):_vm._e(),_vm._v(" "),(mediaItem.type === 'video')?_c('video',{staticClass:"okeReviews-mediaSlider-slide-media okeReviews-mediaSlider-slide-media--video",attrs:{"src":mediaItem.url,"poster":mediaItem.videoPosterURL,"width":"100%","controls":""}},[_c('source',{attrs:{"src":mediaItem.url,"type":"video/mp4"}}),_vm._v("\n                                                Your browser doesn't support HTML5 videos.\n                                            ")]):_vm._e()])}),0)])]),_vm._v(" "),_c('div',{staticClass:"okeReviews-mediaGalleryReviewDialog-review"},[_c('div',{staticClass:"okeReviews-mediaGalleryReviewDialog-review-inner",class:{ 'is-loading': !_vm.review }},[(_vm.review)?_c('okendo-review-simple',{attrs:{"review":_vm.review,"showProductData":_vm.isAllReviews}}):_vm._e()],1)])]),_vm._v(" "),(!_vm.isMediaLoading && _vm.mode === 'aggregate')?_c('div',{ref:"sliderNav",staticClass:"okeReviews-mediaGalleryReviewDialog-nav",class:{ 'is-loading': _vm.isLoading, 'is-loadingMore': _vm.isLoadingMore }},_vm._l((_vm.media),function(mediaItem){return _c('div',{key:mediaItem.thumbnailUrl,ref:mediaItem.append ? 'navSlideToAppend' : undefined,refInFor:true,staticClass:"okeReviews-mediaGalleryReviewDialog-nav-slide"},[(mediaItem.type === 'image')?_c('img',{staticClass:"okeReviews-mediaGalleryReviewDialog-nav-slide-media",attrs:{"src":mediaItem.thumbnailUrl}}):_vm._e(),_vm._v(" "),(mediaItem.type === 'video')?_c('div',{staticClass:"okeReviews-mediaGalleryReviewDialog-nav-slide-media",style:({ 'background-image': ("url('" + (mediaItem.thumbnailUrl) + "')") })},[_c('div',{staticClass:"okeReviews-mediaGalleryReviewDialog-nav-slide-media-playOverlay"})]):_vm._e()])}),0):_vm._e(),_vm._v(" "),(_vm.isError)?_c('div',{staticClass:"okeReviews-mediaGalleryReviewDialog-errorOverlay"},[_c('div',{staticClass:"okeReviews-mediaGalleryReviewDialog-errorOverlay-inner"},[_c('div',{staticClass:"okeReviews-mediaGalleryReviewDialog-errorOverlay-img"}),_vm._v(" "),_c('div',{staticClass:"okeReviews-mediaGalleryReviewDialog-errorOverlay-msg"}),_vm._v(" "),_c('button',{staticClass:"okeReviews-button",on:{"click":function($event){return _vm.loadMediaGalleryData()}}},[_vm._v("Retry")])])]):_vm._e()])]),_vm._v(" "),_c('button',{staticClass:"okeReviews-modal-dialog-close",attrs:{"type":"button","aria-label":"close this modal window"},on:{"click":function($event){return _vm.onClose()}}})])])])])};
var __vue_staticRenderFns__$f = [];

  /* style */
  const __vue_inject_styles__$f = function (inject) {
    if (!inject) return
    inject("data-v-eaf73b2a_0", { source: ".okeReviews .okeReviews-mediaGalleryReviewDialog-nav{position:relative;height:96px;overflow:hidden}.okeReviews .okeReviews-mediaGalleryReviewDialog-slider-inner{max-height:690px}", map: undefined, media: undefined });

  };
  /* scoped */
  const __vue_scope_id__$f = undefined;
  /* module identifier */
  const __vue_module_identifier__$f = undefined;
  /* functional template */
  const __vue_is_functional_template__$f = false;
  /* style inject SSR */
  
  /* style inject shadow dom */
  

  
  var OkendoMediaGallery = normalizeComponent(
    { render: __vue_render__$f, staticRenderFns: __vue_staticRenderFns__$f },
    __vue_inject_styles__$f,
    __vue_script__$f,
    __vue_scope_id__$f,
    __vue_is_functional_template__$f,
    __vue_module_identifier__$f,
    false,
    createInjector,
    undefined,
    undefined
  );

//
// Local dev/serve environment
// Local storage for helpful voting
// --- Skipping for now ---
// Template support (minimal template etc.)
// Foursixty integration
// GA
// Translation

var script$g = {
  props: {
    productId: {
      type: String,
      default: ''
    },
    sizeOverride: {
      type: String
    }
  },
  components: {
    OkendoReview,
    OkendoState,
    OkendoAggregate,
    OkendoMediaGallery,
    OkendoWriteReviewButton
  },

  data() {
    return {
      apiStoreBaseUrl: '',
      shopifyProductId: '',
      writeReviewUrl: '',
      nextUrl: '',
      reviews: [],
      aggregate: undefined,
      mediaAggregate: undefined,
      viewState: 'loading',
      isSorting: false,
      isLoadingMore: false,
      size: 'small',
      isMediaModalOpen: false,
      mediaModalConfig: {}
    };
  },

  mounted() {
    if (!this.sizeOverride) {
      this.addResizeListener();
    }

    this.isAllReviews = !this.productId;
    this.shopifyProductId = OkendoAPI.transformProductId(this.productId);
    this.writeReviewUrl = this.getWriteReviewUrl();
    const reviewsPromise = this.getReviewsPromise();
    const aggregatePromise = this.getAggregatePromise();
    const mediaAggregatePromise = this.getMediaPromise();
    Promise.all([reviewsPromise, aggregatePromise, mediaAggregatePromise]).then(([reviewsResponse, aggregateResponse, mediaResponse]) => {
      if (reviewsResponse && aggregateResponse && mediaResponse) {
        this.aggregate = aggregateResponse.data.reviewAggregate;
        this.mediaAggregate = mediaResponse.data;
        const {
          nextUrl,
          reviews
        } = reviewsResponse.data;
        this.nextUrl = nextUrl;
        this.reviews = reviews;
        this.viewState = reviews.length ? 'content' : 'empty';
      } else {
        this.viewState = 'error';
      }
    }).catch(() => {
      this.viewState = 'error';
    });
  },

  destroyed() {
    this.removeResizeListener();
  },

  computed: {
    widgetSize() {
      return this.sizeOverride || this.size;
    }

  },
  methods: {
    getReviewsPromise(orderBy = 'date desc') {
      return this.isAllReviews ? OkendoAPI.getReviewsAll(undefined, orderBy) : OkendoAPI.getReviewsProduct(this.shopifyProductId, undefined, orderBy);
    },

    getAggregatePromise() {
      return this.isAllReviews ? OkendoAPI.getReviewAggregateAll() : OkendoAPI.getReviewAggregateProduct(this.shopifyProductId);
    },

    getMediaPromise() {
      return this.isAllReviews ? OkendoAPI.getMediaAll() : OkendoAPI.getMediaProduct(this.shopifyProductId);
    },

    getWriteReviewUrl() {
      return this.isAllReviews ? '' : OkendoAPI.writeReviewUrl(this.shopifyProductId);
    },

    loadMoreReviews() {
      if (!this.isLoadingMore) {
        this.isLoadingMore = true;
        OkendoAPI.getMore(this.nextUrl).then(response => {
          if (response && response.data) {
            const {
              nextUrl,
              reviews
            } = response.data;
            this.nextUrl = nextUrl;
            this.reviews.push(...reviews);
          }

          this.isLoadingMore = false;
        }).catch(() => {
          this.isLoadingMore = false;
        });
      }
    },

    onSortChange(event) {
      const newSort = event && event.target && event.target.value;

      if (newSort) {
        this.isSorting = true;
        this.getReviewsPromise(newSort).then(reviewsResponse => {
          const {
            nextUrl,
            reviews
          } = reviewsResponse.data;
          this.reviews = reviews;
          this.nextUrl = nextUrl;
          this.isSorting = false;
        }).catch(() => {
          this.isSorting = false;
          this.viewState = 'error';
        });
      }
    },

    addResizeListener() {
      this.setWidgetSize();
      window.addEventListener('resize', this.setWidgetSize, {
        passive: true
      });
    },

    removeResizeListener() {
      window.removeEventListener('resize', this.setWidgetSize);
    },

    setWidgetSize() {
      const widgetWidth = this.$refs.widget && this.$refs.widget.clientWidth;

      if (widgetWidth > 960) {
        this.size = 'large';
      } else if (widgetWidth > 640) {
        this.size = 'medium';
      } else {
        this.size = 'small';
      }
    },

    openMediaModal(mediaModalMessage) {
      this.mediaModalConfig = mediaModalMessage;
      this.isMediaModalOpen = true;
    }

  }
};

/* script */
const __vue_script__$g = script$g;

/* template */
var __vue_render__$g = function () {var _vm=this;var _h=_vm.$createElement;var _c=_vm._self._c||_h;return _c('div',{staticClass:"okeReviews okeReviews--theme"},[_c('div',{ref:"widget",staticClass:"okeReviews-reviewsWidget js-okeReviews-reviewsWidget",class:_vm.widgetSize !== 'small' ? ("is-okeReviews-reviewsWidget-" + _vm.widgetSize) : ''},[(_vm.viewState !== 'content')?_c('okendo-state',{attrs:{"viewState":_vm.viewState,"writeReviewUrl":_vm.writeReviewUrl}}):_vm._e(),_vm._v(" "),(!_vm.isLoadingInitial && !_vm.isError && _vm.reviews.length)?[_c('div',{staticClass:"okeReviews-reviewsWidget-header"},[_vm._m(0),_vm._v(" "),_c('div',{staticClass:"okeReviews-reviewsWidget-header-controls"},[_c('okendo-write-review-button',{attrs:{"url":_vm.writeReviewUrl}})],1)]),_vm._v(" "),_c('okendo-aggregate',{attrs:{"aggregate":_vm.aggregate,"mediaAggregate":_vm.mediaAggregate},on:{"mediaClick":function($event){return _vm.openMediaModal($event)}}}),_vm._v(" "),_c('div',{staticClass:"okeReviews-reviewsWidget-reviews"},[_c('div',{staticClass:"okeReviews-reviews"},[_c('div',{staticClass:"okeReviews-reviews-controls"},[_c('div',{staticClass:"okeReviews-reviews-controls-reviewCount"},[_vm._v("\n                            "+_vm._s(_vm.aggregate.reviewCount || '0')+" "+_vm._s(_vm.aggregate.reviewCount === 1 ? 'Review' : 'Reviews')+"\n                        ")]),_vm._v(" "),_c('div',{staticClass:"okeReviews-reviews-controls-select"},[_c('span',{staticClass:"okeReviews-select"},[_c('label',{staticClass:"okeReviews-a11yText",attrs:{"for":"okeReviews-reviews-controls-sort"}},[_vm._v("Sort By")]),_vm._v(" "),_c('select',{staticClass:"okeReviews-select-input",attrs:{"id":"okeReviews-reviews-controls-sort"},on:{"change":function($event){return _vm.onSortChange($event)}}},[_c('option',{attrs:{"value":"date desc","selected":""}},[_vm._v("Most Recent")]),_vm._v(" "),_c('option',{attrs:{"value":"date asc"}},[_vm._v("Oldest")]),_vm._v(" "),_c('option',{attrs:{"value":"has_media desc"}},[_vm._v("Photos & Videos")]),_vm._v(" "),_c('option',{attrs:{"value":"rating desc"}},[_vm._v("Highest Rating")]),_vm._v(" "),_c('option',{attrs:{"value":"rating asc"}},[_vm._v("Lowest Rating")]),_vm._v(" "),_c('option',{attrs:{"value":"helpful desc"}},[_vm._v("Most Helpful")]),_vm._v(" "),_c('option',{attrs:{"value":"unhelpful desc"}},[_vm._v("Least Helpful")])])])])]),_vm._v(" "),_c('div',{staticClass:"okeReviews-alert okeReviews-alert--top",class:{ 'is-okeReviews-visible' : _vm.isSorting },attrs:{"aria-hidden":_vm.isSorting ? 'false' : 'true',"role":_vm.isSorting ? 'alert' : 'none'}},[_c('div',{staticClass:"okeReviews-reviews-alert-inner okeReviews-alert-inner"},[_vm._v("\n                            Loading Reviews...\n                        ")])]),_vm._v(" "),_c('div',{staticClass:"okeReviews-reviews-main"},_vm._l((_vm.reviews),function(review){return _c('div',{key:review.reviewId,staticClass:"okeReviews-reviews-review"},[_c('okendo-review',{attrs:{"review":review,"showProductData":_vm.isAllReviews},on:{"mediaClick":function($event){return _vm.openMediaModal($event)}}})],1)}),0)])]),_vm._v(" "),(_vm.nextUrl)?_c('div',{staticClass:"okeReviews-reviews-footer"},[_c('button',{staticClass:"okeReviews-showMore",class:{ 'is-okeReviews-loading': _vm.isLoadingMore },attrs:{"disabled":_vm.isLoadingMore},on:{"click":function($event){return _vm.loadMoreReviews()}}},[_vm._v("\n                    "+_vm._s(_vm.isLoadingMore ? 'Loading...' : 'Show more')+"\n                ")])]):_vm._e()]:_vm._e()],2),_vm._v(" "),(_vm.isMediaModalOpen)?_c('okendo-media-gallery',{attrs:{"initialIndex":_vm.mediaModalConfig.initialIndex,"mediaItems":_vm.mediaModalConfig.mediaItems,"initialReviewId":_vm.mediaModalConfig.reviewId,"shopifyProductId":_vm.shopifyProductId,"mode":_vm.mediaModalConfig.source,"isAllReviews":_vm.isAllReviews},on:{"close":function($event){_vm.isMediaModalOpen = false;}}}):_vm._e()],1)};
var __vue_staticRenderFns__$g = [function () {var _vm=this;var _h=_vm.$createElement;var _c=_vm._self._c||_h;return _c('div',{staticClass:"okeReviews-reviewsWidget-header-poweredBy"},[_c('div',{staticClass:"okeReviews-reviewsWidget-poweredBy"},[_c('a',{staticClass:"okeReviews-poweredBy",attrs:{"href":"https://www.okendo.io","target":"_blank","aria-label":"Powered by Okendo"}})])])}];

  /* style */
  const __vue_inject_styles__$g = function (inject) {
    if (!inject) return
    inject("data-v-1848aaca_0", { source: ".okeReviews .okeReviews-reviewsWidget{padding-top:0}.okeReviews .okeReviews-reviewsWidget-poweredBy{display:block}.okeReviews .okeReviews-reviewsWidget-header{display:flex;align-items:center}.okeReviews .modal-wrapper{border:0}.okeReviews .okeReviews-showMore{margin-top:-2px}.okeReviews .okeReviews-showMore::after{display:none}.okeReviews .okeReviews-showMore.is-okeReviews-loading::after{display:inline-block;animation:spin 1s linear infinite;border:2px solid;border-color:currentColor currentColor transparent transparent;border-radius:50%;height:10px;width:10px}.okeReviews .okeReviews-review-primary,.okeReviews .okeReviews-reviewsAggregate-primary{border:2px solid #f5f5f5!important}.okeReviews .okeReviews-review-side,.okeReviews .okeReviews-reviewsAggregate-side{border:0!important;background-color:#f5f5f5}.okeReviews .okeReviews-mediaStrip-item-media{border:2px solid #efefef}.okeReviews .okeReviews-showMore{border:0;padding-left:24px;padding-right:24px;background-color:#f5f5f5}.okeReviews .okeReviews-showMore{position:relative}.okeReviews .okeReviews-reviews-footer::before{height:2px;background:#f5f5f5;top:-2px}.okeReviews .okeReviews-review-reviewer-profile{border-bottom:2px solid #efefef}.okeReviews .okeReviews-review-reviewer-profile-details{vertical-align:middle;padding-top:0}.okeReviews .okeReviews-avatar{background-color:#fff;width:50px;height:50px}.okeReviews .okeReviews-avatar-placeholder{font-size:1em}.okeReviews .okeReviews-avatar::after{background-size:15px;width:15px;height:15px}.okeReviews .okeReviews-review--simple .okeReviews-avatar{background-color:#efefef}.okeReviews .okeReviews-helpful-vote-button,.okeReviews .okeReviews-review-recommendation,.okeReviews .okeReviews-review-reviewer-profile-status{margin-top:0;vertical-align:baseline}.okeReviews .okeReviews-helpful-vote-button::before,.okeReviews .okeReviews-review-recommendation::before,.okeReviews .okeReviews-review-reviewer-profile-status::before{margin-top:-.4em;margin-right:.1em;font-size:1em}.okeReviews .okeReviews-select,.okeReviews .okeReviews-select::before{border-color:#f5f5f5;border-width:2px}.okeReviews .okeReviews-reviewsWidget .oke-helpful{text-align:right}@keyframes spin{from{transform:rotate(0)}to{transform:rotate(360deg)}}", map: undefined, media: undefined });

  };
  /* scoped */
  const __vue_scope_id__$g = undefined;
  /* module identifier */
  const __vue_module_identifier__$g = undefined;
  /* functional template */
  const __vue_is_functional_template__$g = false;
  /* style inject SSR */
  
  /* style inject shadow dom */
  

  
  var ReviewsWidget = normalizeComponent(
    { render: __vue_render__$g, staticRenderFns: __vue_staticRenderFns__$g },
    __vue_inject_styles__$g,
    __vue_script__$g,
    __vue_scope_id__$g,
    __vue_is_functional_template__$g,
    __vue_module_identifier__$g,
    false,
    createInjector,
    undefined,
    undefined
  );

//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
var script$h = {
  props: {
    avatarUrl: {
      type: String,
      required: false
    },
    displayName: {
      type: String,
      required: true
    },
    width: {
      type: String,
      required: false,
      default: '40px'
    }
  },
  computed: {
    initials() {
      return (this.displayName || '').split(' ').slice(0, 2).map(part => part.substr(0, 1).toLocaleUpperCase()).join('');
    }

  }
};

/* script */
const __vue_script__$h = script$h;

/* template */
var __vue_render__$h = function () {var _vm=this;var _h=_vm.$createElement;var _c=_vm._self._c||_h;return _c('div',{staticClass:"oke-avatar",style:({ width: _vm.width, height: _vm.width, 'line-height': _vm.width })},[_c('span',{staticClass:"oke-avatar-placeholder"},[_vm._v("\n        "+_vm._s(_vm.initials)+"\n    ")]),_vm._v(" "),(_vm.avatarUrl)?_c('img',{staticClass:"oke-avatar-image",attrs:{"src":_vm.avatarUrl,"alt":""}}):_vm._e()])};
var __vue_staticRenderFns__$h = [];

  /* style */
  const __vue_inject_styles__$h = function (inject) {
    if (!inject) return
    inject("data-v-14c76f53_0", { source: ".oke-avatar{background-color:#eee;width:32px;height:32px;margin-right:16px;line-height:32px;border-radius:50%;overflow:hidden;position:relative}.oke-avatar-image{display:block;position:absolute;width:100%;height:100%}.oke-avatar-placeholder{font-size:12px;font-weight:600;user-select:none;width:100%;left:0;position:absolute;text-align:center}", map: undefined, media: undefined });

  };
  /* scoped */
  const __vue_scope_id__$h = undefined;
  /* module identifier */
  const __vue_module_identifier__$h = undefined;
  /* functional template */
  const __vue_is_functional_template__$h = false;
  /* style inject SSR */
  
  /* style inject shadow dom */
  

  
  var Avatar = normalizeComponent(
    { render: __vue_render__$h, staticRenderFns: __vue_staticRenderFns__$h },
    __vue_inject_styles__$h,
    __vue_script__$h,
    __vue_scope_id__$h,
    __vue_is_functional_template__$h,
    __vue_module_identifier__$h,
    false,
    createInjector,
    undefined,
    undefined
  );

//
var script$i = {
  props: {
    question: {
      type: Object,
      required: true
    },
    index: {
      type: Number,
      required: true
    }
  },
  components: {
    Avatar,
    HelpfulVoting
  },
  methods: {
    formatDate(date) {
      return OkendoUtils.getPrettyDate(date);
    }

  }
};

/* script */
const __vue_script__$i = script$i;

/* template */
var __vue_render__$i = function () {var _vm=this;var _h=_vm.$createElement;var _c=_vm._self._c||_h;return (_vm.question)?_c('div',[_c('div',{staticClass:"oke-qw-question"},[_c('div',{staticClass:"oke-qw-question-avatar"},[_vm._m(0)],1),_vm._v(" "),_c('div',{staticClass:"oke-qw-question-main"},[_vm._m(1),_vm._v(" "),_vm._m(2),_vm._v(" "),_vm._m(3),_vm._v(" "),_c('div',{staticClass:"oke-qw-question-main-helpful"},[_c('helpful-voting',{attrs:{"itemId":_vm.question.questionId,"helpfulCount":_vm.question.helpfulCount,"unhelpfulCount":_vm.question.unhelpfulCount,"contentType":"question"}})],1)])]),_vm._v(" "),_vm._l((_vm.question.answers),function(answer){return _c('div',{key:answer.answerId,staticClass:"oke-qw-answer"},[_c('div',{staticClass:"oke-qw-answer-avatar"},[_vm._o(_c('avatar',{attrs:{"avatarUrl":answer.author.avatarUrl,"displayName":answer.author.displayName,"width":"40px"}}),0,answer.answerId)],1),_vm._v(" "),_c('div',{staticClass:"oke-qw-answer-main"},[_vm._o(_c('div',{staticClass:"oke-qw-answer-main-author"},[_vm._v("\n                "+_vm._s(answer.author.displayName)+"\n            ")]),1,answer.answerId),_vm._v(" "),_vm._o(_c('div',{staticClass:"oke-qw-answer-main-date"},[_vm._v("\n                "+_vm._s(_vm.formatDate(answer.dateCreated))+"\n            ")]),2,answer.answerId),_vm._v(" "),_vm._o(_c('div',{staticClass:"oke-qw-answer-main-body",domProps:{"innerHTML":_vm._s(answer.body)}}),3,answer.answerId),_vm._v(" "),_c('div',{staticClass:"oke-qw-answer-main-helpful"},[_c('helpful-voting',{attrs:{"itemId":answer.answerId,"parentId":_vm.question.questionId,"helpfulCount":answer.helpfulCount,"unhelpfulCount":answer.unhelpfulCount,"contentType":"answer"}})],1)])])})],2):_vm._e()};
var __vue_staticRenderFns__$i = [function () {var _vm=this;var _h=_vm.$createElement;var _c=_vm._self._c||_h;return _c('avatar',{attrs:{"avatarUrl":_vm.question.author.avatarUrl,"displayName":_vm.question.author.displayName,"width":"40px"}})},function () {var _vm=this;var _h=_vm.$createElement;var _c=_vm._self._c||_h;return _c('div',{staticClass:"oke-qw-question-main-author"},[_vm._v("\n                "+_vm._s(_vm.question.author.displayName)+"\n            ")])},function () {var _vm=this;var _h=_vm.$createElement;var _c=_vm._self._c||_h;return _c('div',{staticClass:"oke-qw-question-main-date"},[_vm._v("\n                "+_vm._s(_vm.formatDate(_vm.question.dateCreated))+"\n            ")])},function () {var _vm=this;var _h=_vm.$createElement;var _c=_vm._self._c||_h;return _c('div',{staticClass:"oke-qw-question-main-body"},[_vm._v("\n                "+_vm._s(_vm.question.body)+"\n            ")])}];

  /* style */
  const __vue_inject_styles__$i = function (inject) {
    if (!inject) return
    inject("data-v-2ce09edd_0", { source: ".oke-qw-answer,.oke-qw-question{display:flex;flex-flow:row nowrap;justify-content:stretch}.oke-qw-answer-main,.oke-qw-question-main{flex:1 1 0}.oke-qw-answer-main-author,.oke-qw-answer-main-date,.oke-qw-question-main-author,.oke-qw-question-main-date{margin-bottom:4px}.oke-qw-answer-main-author,.oke-qw-question-main-author{font-weight:700;font-size:12px;float:left}.oke-qw-answer-main-body,.oke-qw-question-main-body{clear:both;margin-bottom:4px}.oke-qw-answer-main-date,.oke-qw-question-main-date{font-size:12px;color:#7c7c7c;float:right}.oke-qw-question{margin-bottom:24px}.oke-qw-question-main{padding-right:20px}.oke-qw-question-main-body{font-weight:700}.oke-qw-answer{margin-left:56px;background-color:#f5f5f5;padding:20px;border-radius:8px;border-radius:12px}.oke-is-small .oke-qw-answer{margin-left:0}", map: undefined, media: undefined });

  };
  /* scoped */
  const __vue_scope_id__$i = undefined;
  /* module identifier */
  const __vue_module_identifier__$i = undefined;
  /* functional template */
  const __vue_is_functional_template__$i = false;
  /* style inject SSR */
  
  /* style inject shadow dom */
  

  
  var Question = normalizeComponent(
    { render: __vue_render__$i, staticRenderFns: __vue_staticRenderFns__$i },
    __vue_inject_styles__$i,
    __vue_script__$i,
    __vue_scope_id__$i,
    __vue_is_functional_template__$i,
    __vue_module_identifier__$i,
    false,
    createInjector,
    undefined,
    undefined
  );

//
var script$j = {
  props: {
    questions: {
      type: Array,
      required: true
    },
    isLoading: {
      type: Boolean,
      required: true
    }
  },
  components: {
    Question
  }
};

/* script */
const __vue_script__$j = script$j;

/* template */
var __vue_render__$j = function () {var _vm=this;var _h=_vm.$createElement;var _c=_vm._self._c||_h;return (_vm.questions.length)?_c('ul',{staticClass:"oke-qw-questionList",class:{ 'oke-is-loading': _vm.isLoading }},_vm._l((_vm.questions),function(question,index){return _c('li',{key:question.questionId,staticClass:"oke-qw-questionList-item"},[_c('question',{attrs:{"question":question,"index":index}})],1)}),0):_vm._e()};
var __vue_staticRenderFns__$j = [];

  /* style */
  const __vue_inject_styles__$j = function (inject) {
    if (!inject) return
    inject("data-v-53594192_0", { source: ".oke-qw-questionList{transition:.2s opacity ease-out;position:relative}.oke-qw-questionList-item{list-style-type:none;margin-bottom:24px;padding-bottom:24px;border-bottom:2px solid #f5f5f5}.oke-qw-questionList-item:last-child{margin-bottom:0;border-bottom:0}.oke-qw-questionList.oke-is-loading{opacity:.3}", map: undefined, media: undefined });

  };
  /* scoped */
  const __vue_scope_id__$j = undefined;
  /* module identifier */
  const __vue_module_identifier__$j = undefined;
  /* functional template */
  const __vue_is_functional_template__$j = false;
  /* style inject SSR */
  
  /* style inject shadow dom */
  

  
  var QuestionsList = normalizeComponent(
    { render: __vue_render__$j, staticRenderFns: __vue_staticRenderFns__$j },
    __vue_inject_styles__$j,
    __vue_script__$j,
    __vue_scope_id__$j,
    __vue_is_functional_template__$j,
    __vue_module_identifier__$j,
    false,
    createInjector,
    undefined,
    undefined
  );

//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
var script$k = {
  props: {
    value: {
      type: String,
      required: false,
      default: 'date desc'
    },
    contentType: {
      type: 'reviews' | 'questions',
      required: false,
      default: 'reviews'
    },
    disabled: {
      type: Boolean,
      required: false,
      default: false
    }
  },
  methods: {
    onSortChange(event) {
      this.$emit('input', event.target.value);
    }

  }
};

/* script */
const __vue_script__$k = script$k;

/* template */
var __vue_render__$k = function () {var _vm=this;var _h=_vm.$createElement;var _c=_vm._self._c||_h;return _c('div',{staticClass:"oke-sortSelect"},[_c('label',{staticClass:"oke-sortSelect-label",attrs:{"for":"oke-sortSelect-sort"}},[_vm._v("\n        Sort\n    ")]),_vm._v(" "),_c('select',{staticClass:"oke-sortSelect-sort oke-select",attrs:{"disabled":_vm.disabled,"id":"oke-sortSelect-sort"},domProps:{"value":_vm.value},on:{"change":function($event){return _vm.onSortChange($event)}}},[_c('option',{attrs:{"value":"date desc"}},[_vm._v(_vm._s('Most Recent'))]),_vm._v(" "),_c('option',{attrs:{"value":"date asc"}},[_vm._v(_vm._s('Oldest'))]),_vm._v(" "),_c('option',{attrs:{"value":"helpful desc"}},[_vm._v(_vm._s('Most Helpful'))]),_vm._v(" "),(_vm.contentType === 'reviews')?[_c('option',{attrs:{"value":"has_media desc"}},[_vm._v(_vm._s('Photos & Videos'))]),_vm._v(" "),_c('option',{attrs:{"value":"rating desc"}},[_vm._v(_vm._s('Highest Rating'))]),_vm._v(" "),_c('option',{attrs:{"value":"rating asc"}},[_vm._v(_vm._s('Lowest Rating'))]),_vm._v(" "),_c('option',{attrs:{"value":"unhelpful desc"}},[_vm._v(_vm._s('Least Helpful'))])]:_vm._e()],2)])};
var __vue_staticRenderFns__$k = [];

  /* style */
  const __vue_inject_styles__$k = function (inject) {
    if (!inject) return
    inject("data-v-e624547e_0", { source: ".oke-sortSelect-label{font-weight:600;margin-right:16px}", map: undefined, media: undefined });

  };
  /* scoped */
  const __vue_scope_id__$k = undefined;
  /* module identifier */
  const __vue_module_identifier__$k = undefined;
  /* functional template */
  const __vue_is_functional_template__$k = false;
  /* style inject SSR */
  
  /* style inject shadow dom */
  

  
  var SortSelect = normalizeComponent(
    { render: __vue_render__$k, staticRenderFns: __vue_staticRenderFns__$k },
    __vue_inject_styles__$k,
    __vue_script__$k,
    __vue_scope_id__$k,
    __vue_is_functional_template__$k,
    __vue_module_identifier__$k,
    false,
    createInjector,
    undefined,
    undefined
  );

//
//
//
//
var script$l = {
  props: {
    size: {
      type: 'small' | 'medium' | 'large',
      required: false,
      default: 'small'
    }
  }
};

/* script */
const __vue_script__$l = script$l;

/* template */
var __vue_render__$l = function () {var _vm=this;var _h=_vm.$createElement;var _c=_vm._self._c||_h;return _c('a',{staticClass:"okeReviews-poweredBy",attrs:{"href":"https://www.okendo.io","target":"_blank","aria-label":"Powered by Okendo"}})};
var __vue_staticRenderFns__$l = [];

  /* style */
  const __vue_inject_styles__$l = function (inject) {
    if (!inject) return
    inject("data-v-39e40518_0", { source: ".oke-poweredBy{display:block;transition:.2s opacity ease-out;opacity:.4}.oke-poweredBy:focus,.oke-poweredBy:hover{outline:0;opacity:.6}.oke-poweredBy-image{display:block}", map: undefined, media: undefined });

  };
  /* scoped */
  const __vue_scope_id__$l = undefined;
  /* module identifier */
  const __vue_module_identifier__$l = undefined;
  /* functional template */
  const __vue_is_functional_template__$l = false;
  /* style inject SSR */
  
  /* style inject shadow dom */
  

  
  var PoweredBy = normalizeComponent(
    { render: __vue_render__$l, staticRenderFns: __vue_staticRenderFns__$l },
    __vue_inject_styles__$l,
    __vue_script__$l,
    __vue_scope_id__$l,
    __vue_is_functional_template__$l,
    __vue_module_identifier__$l,
    false,
    createInjector,
    undefined,
    undefined
  );

var script$m = {
    props: {
        productId: {
            type: String,
            required: false
        }
    },
    computed: {
        askQuestionUrl() {
            return OkendoAPI.askQuestionUrl(this.productId);
        }
    }
};

/* script */
const __vue_script__$m = script$m;

/* template */
var __vue_render__$m = function () {var _vm=this;var _h=_vm.$createElement;var _c=_vm._self._c||_h;return (_vm.productId)?_vm._m(0):_vm._e()};
var __vue_staticRenderFns__$m = [function () {var _vm=this;var _h=_vm.$createElement;var _c=_vm._self._c||_h;return _c('a',{staticClass:"oke-qw-askQuestion oke-button",attrs:{"href":_vm.askQuestionUrl,"target":"_blank","rel":"noopener noreferrer"}},[_vm._v("\n    Ask a Question\n")])}];

  /* style */
  const __vue_inject_styles__$m = undefined;
  /* scoped */
  const __vue_scope_id__$m = undefined;
  /* module identifier */
  const __vue_module_identifier__$m = undefined;
  /* functional template */
  const __vue_is_functional_template__$m = false;
  /* style inject */
  
  /* style inject SSR */
  
  /* style inject shadow dom */
  

  
  var AskQuestionButton = normalizeComponent(
    { render: __vue_render__$m, staticRenderFns: __vue_staticRenderFns__$m },
    __vue_inject_styles__$m,
    __vue_script__$m,
    __vue_scope_id__$m,
    __vue_is_functional_template__$m,
    __vue_module_identifier__$m,
    false,
    undefined,
    undefined,
    undefined
  );

//
var script$n = {
  props: {
    initialSortOption: {
      type: String,
      required: false,
      default: 'date desc'
    },
    widgetSize: {
      type: 'small' | 'medium' | 'large',
      default: 'small',
      required: false
    },
    sortDisabled: {
      type: Boolean,
      default: false,
      required: false
    },
    productId: {
      type: String,
      required: true
    }
  },
  components: {
    AskQuestionButton,
    PoweredBy,
    SortSelect
  },

  data() {
    return {
      selectedSortOption: 'date desc'
    };
  },

  mounted() {
    this.selectedSortOption = this.initialSortOption;
  },

  methods: {
    onSortChange(sortOption) {
      this.$emit('sortChange', sortOption);
    }

  }
};

/* script */
const __vue_script__$n = script$n;

/* template */
var __vue_render__$n = function () {var _vm=this;var _h=_vm.$createElement;var _c=_vm._self._c||_h;return _c('div',{staticClass:"oke-qw-header"},[_c('div',{staticClass:"oke-qw-header-poweredBy"},[_c('powered-by',{attrs:{"size":_vm.widgetSize}})],1),_vm._v(" "),_c('div',{staticClass:"oke-qw-header-sort"},[_c('sort-select',{attrs:{"disabled":_vm.sortDisabled,"contentType":"questions"},on:{"input":function($event){return _vm.onSortChange($event)}},model:{value:(_vm.selectedSortOption),callback:function ($$v) {_vm.selectedSortOption=$$v;},expression:"selectedSortOption"}})],1),_vm._v(" "),_c('div',{staticClass:"oke-qw-header-askQuestion"},[_c('ask-question-button',{attrs:{"productId":_vm.productId}})],1)])};
var __vue_staticRenderFns__$n = [];

  /* style */
  const __vue_inject_styles__$n = function (inject) {
    if (!inject) return
    inject("data-v-c4da3d8c_0", { source: ".oke-qw-header{display:flex;align-items:center;padding-bottom:12px}.oke-qw-header-sort{margin-left:auto;margin-right:24px}.oke-is-small .oke-qw-header{flex-flow:row wrap;justify-content:space-between}.oke-is-small .oke-qw-header-askQuestion{flex-basis:100%}.oke-is-small .oke-qw-header-askQuestion .oke-qw-askQuestion{width:100%;margin-top:8px}.oke-is-small .oke-qw-header-sort{margin:0}", map: undefined, media: undefined });

  };
  /* scoped */
  const __vue_scope_id__$n = undefined;
  /* module identifier */
  const __vue_module_identifier__$n = undefined;
  /* functional template */
  const __vue_is_functional_template__$n = false;
  /* style inject SSR */
  
  /* style inject shadow dom */
  

  
  var QuestionsHeader = normalizeComponent(
    { render: __vue_render__$n, staticRenderFns: __vue_staticRenderFns__$n },
    __vue_inject_styles__$n,
    __vue_script__$n,
    __vue_scope_id__$n,
    __vue_is_functional_template__$n,
    __vue_module_identifier__$n,
    false,
    createInjector,
    undefined,
    undefined
  );

//
//
//
//
//
//
//
//
//
//
//
//
//
//
var script$o = {
  props: {
    isLoading: {
      type: Boolean,
      default: false
    }
  },
  methods: {
    onClick(event) {
      this.$emit('click', event.target.value);
    }

  }
};

/* script */
const __vue_script__$o = script$o;

/* template */
var __vue_render__$o = function () {var _vm=this;var _h=_vm.$createElement;var _c=_vm._self._c||_h;return _c('div',{staticClass:"oke-showMore oke-is-loading"},[_c('button',{staticClass:"oke-showMore-button oke-button",class:{ 'oke-is-loading': _vm.isLoading },attrs:{"title":_vm.isLoading ? 'Loading more' : '',"type":"button"},on:{"click":function($event){return _vm.onClick($event)}}},[_c('span',{staticClass:"oke-showMore-button-text"},[_vm._v("Show More")])])])};
var __vue_staticRenderFns__$o = [];

  /* style */
  const __vue_inject_styles__$o = function (inject) {
    if (!inject) return
    inject("data-v-fa41b2e6_0", { source: ".oke-showMore{text-align:center}.oke-showMore-button{border-top:0!important}.oke-showMore-button.oke-is-loading .oke-showMore-button-text{opacity:0}.oke-showMore-button-text{font-family:inherit}", map: undefined, media: undefined });

  };
  /* scoped */
  const __vue_scope_id__$o = undefined;
  /* module identifier */
  const __vue_module_identifier__$o = undefined;
  /* functional template */
  const __vue_is_functional_template__$o = false;
  /* style inject SSR */
  
  /* style inject shadow dom */
  

  
  var ShowMore = normalizeComponent(
    { render: __vue_render__$o, staticRenderFns: __vue_staticRenderFns__$o },
    __vue_inject_styles__$o,
    __vue_script__$o,
    __vue_scope_id__$o,
    __vue_is_functional_template__$o,
    __vue_module_identifier__$o,
    false,
    createInjector,
    undefined,
    undefined
  );

//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
var script$p = {
  props: {
    position: {
      type: 'top' | 'bottom',
      required: true
    }
  },

  data() {
    return {
      isVisible: false,
      message: 'Loading...'
    };
  },

  methods: {
    show(message, timeout = 0) {
      this.isVisible = true;
      this.message = message;

      if (timeout) {
        setTimeout(() => {
          this.isVisible = false;
        }, timeout);
      }
    },

    hide() {
      this.isVisible = false;
    }

  }
};

/* script */
const __vue_script__$p = script$p;

/* template */
var __vue_render__$p = function () {var _vm=this;var _h=_vm.$createElement;var _c=_vm._self._c||_h;return _c('div',{staticClass:"oke-alert",class:{
        'oke-is-visible': _vm.isVisible && _vm.message.length,
        'oke-alert--top': _vm.position === 'top',
        'oke-alert--bottom': _vm.position === 'bottom'
    },attrs:{"aria-hidden":!_vm.isVisible,"role":"alert","aria-live":"assertive","aria-atomic":"true"}},[_c('div',{staticClass:"oke-alert-inner"},[_vm._v("\n        "+_vm._s(_vm.message)+"\n    ")])])};
var __vue_staticRenderFns__$p = [];

  /* style */
  const __vue_inject_styles__$p = function (inject) {
    if (!inject) return
    inject("data-v-199817f3_0", { source: ".oke-alert{position:absolute;overflow:hidden;padding:12px 0;left:0;right:0;margin-left:auto;margin-right:auto;text-align:center}.oke-alert-inner{display:inline-block;transition:transform .2s ease-out;background-color:#fff;padding:8px 24px;box-shadow:0 0 5px 0 rgba(0,0,0,.15)}.oke-alert--top{top:0}.oke-alert--top .oke-alert-inner{top:12px;transform:translateY(-150%)}.oke-alert--bottom{bottom:0}.oke-alert--bottom .oke-alert-inner{bottom:12px;transform:translateY(150%)}.oke-alert.oke-is-visible.oke-alert--bottom .oke-alert-inner,.oke-alert.oke-is-visible.oke-alert--top .oke-alert-inner{top:auto;transform:translateY(0)}", map: undefined, media: undefined });

  };
  /* scoped */
  const __vue_scope_id__$p = undefined;
  /* module identifier */
  const __vue_module_identifier__$p = undefined;
  /* functional template */
  const __vue_is_functional_template__$p = false;
  /* style inject SSR */
  
  /* style inject shadow dom */
  

  
  var Alert = normalizeComponent(
    { render: __vue_render__$p, staticRenderFns: __vue_staticRenderFns__$p },
    __vue_inject_styles__$p,
    __vue_script__$p,
    __vue_scope_id__$p,
    __vue_is_functional_template__$p,
    __vue_module_identifier__$p,
    false,
    createInjector,
    undefined,
    undefined
  );

//
var script$q = {
  props: {
    productId: {
      type: String,
      required: true
    },
    sizeOverride: {
      type: String
    }
  },
  components: {
    Alert,
    AskQuestionButton,
    QuestionsHeader,
    QuestionsList,
    ShowMore
  },

  data() {
    return {
      viewState: 'loading',
      size: 'small',
      questions: [],
      isLoadingMore: false,
      selectedSortOption: 'date desc',
      nextUrl: '',
      loadInitialLimit: 5,
      shopifyProductId: ''
    };
  },

  mounted() {
    this.shopifyProductId = OkendoAPI.transformProductId(this.productId);

    if (!this.sizeOverride) {
      this.addResizeListener();
    }

    this.loadQuestions(this.loadInitialLimit, this.selectedSortOption);
  },

  destroyed() {
    this.removeResizeListener();
  },

  computed: {
    widgetSize() {
      return this.sizeOverride || this.size;
    }

  },
  methods: {
    loadQuestions(limit, sort) {
      if (this.shopifyProductId) {
        const topAlert = this.$refs.topAlert;

        if (topAlert) {
          topAlert.show('Loading Questions...');
        }

        OkendoAPI.getQuestionsProduct(this.shopifyProductId, limit, sort).then(({
          data
        }) => {
          this.questions = data.questions;
          this.nextUrl = data.nextUrl;
          this.viewState = this.questions.length ? 'content' : 'empty';

          if (topAlert) {
            topAlert.hide();
          }
        }).catch(() => {
          this.viewState = 'error';

          if (topAlert) {
            topAlert.hide();
          }
        });
      }
    },

    showMoreQuestions() {
      if (this.nextUrl) {
        this.isLoadingMore = true;
        OkendoAPI.getMore(this.nextUrl).then(({
          data
        }) => {
          this.questions = [...this.questions, ...data.questions];
          this.nextUrl = data.nextUrl;
          this.isLoadingMore = false;
        }).catch(() => {
          this.$refs.bottomAlert.show('Unexpected error occurred, please try again', 2000);
          this.isLoadingMore = false;
        });
      }
    },

    onSortChange(newSort) {
      this.selectedSortOption = newSort;
      this.loadQuestions(this.loadInitialLimit, this.selectedSortOption);
    },

    addResizeListener() {
      this.setWidgetSize();
      window.addEventListener('resize', this.setWidgetSize, {
        passive: true
      });
    },

    removeResizeListener() {
      window.removeEventListener('resize', this.setWidgetSize);
    },

    setWidgetSize() {
      const widgetWidth = this.$refs.questionsWidget && this.$refs.questionsWidget.clientWidth;

      if (widgetWidth > 960) {
        this.size = 'large';
      } else if (widgetWidth > 640) {
        this.size = 'medium';
      } else {
        this.size = 'small';
      }
    }

  }
};

/* script */
const __vue_script__$q = script$q;

/* template */
var __vue_render__$q = function () {var _vm=this;var _h=_vm.$createElement;var _c=_vm._self._c||_h;return _c('div',{ref:"questionsWidget",staticClass:"okeReviews",attrs:{"data-oke-reviews-version":""}},[_c('div',{staticClass:"okeReviews-qandaWidget oke-qw",class:{
            'oke-is-small is-okeReviews-qandaWidget-small': _vm.widgetSize === 'small',
            'oke-is-medium is-okeReviews-qandaWidget-medium': _vm.widgetSize === 'medium',
            'oke-is-large is-okeReviews-qandaWidget-large': _vm.widgetSize === 'large'
        }},[(_vm.viewState === 'loading' || _vm.viewState === 'content')?_c('questions-header',{attrs:{"sortDisabled":_vm.viewState !== 'content',"productId":_vm.shopifyProductId,"selectedSortOption":_vm.selectedSortOption,"widgetSize":_vm.widgetSize},on:{"sortChange":function($event){return _vm.onSortChange($event)}}}):_vm._e(),_vm._v(" "),_c('div',{staticClass:"oke-qw-main"},[(_vm.viewState === 'loading' || _vm.viewState === 'content')?_c('questions-list',{attrs:{"isLoading":_vm.viewState === 'loading',"questions":_vm.questions}}):_vm._e(),_vm._v(" "),(_vm.viewState === 'loading')?_c('div',{staticClass:"oke-qw-loading"},[_vm._m(0)]):_vm._e(),_vm._v(" "),(_vm.viewState === 'empty')?_c('div',{staticClass:"oke-qw-empty"},[_c('div',{staticClass:"oke-qw-empty-text"},[_vm._v("\n                    No questions yet, ask one now?\n                ")]),_vm._v(" "),_c('div',{staticClass:"oke-qw-empty-button"},[_c('ask-question-button',{attrs:{"productId":_vm.shopifyProductId}})],1)]):_vm._e(),_vm._v(" "),(_vm.viewState === 'error')?_c('div',{staticClass:"oke-qw-error"},[_c('div',{staticClass:"oke-qw-error-text"},[_vm._v("\n                    Error occurred while loading questions 😣\n                ")]),_vm._v(" "),_c('div',{staticClass:"oke-qw-error-button"},[_c('button',{staticClass:"oke-qw-error-button oke-button",attrs:{"type":"button"},on:{"click":function($event){return _vm.loadQuestions(_vm.loadInitialLimit, _vm.selectedSortOption)}}},[_vm._v("\n                        Retry\n                    ")])])]):_vm._e(),_vm._v(" "),(_vm.questions && _vm.questions.length)?_c('div',{staticClass:"oke-qw-main-alerts"},[_c('alert',{ref:"bottomAlert",attrs:{"position":"bottom"}}),_vm._v(" "),_c('alert',{ref:"topAlert",attrs:{"position":"top"}})],1):_vm._e()],1),_vm._v(" "),_c('div',{staticClass:"oke-qw-footer"},[(_vm.nextUrl && _vm.viewState === 'content')?_c('show-more',{attrs:{"isLoading":_vm.isLoadingMore},on:{"click":function($event){return _vm.showMoreQuestions()}}}):_vm._e()],1)],1)])};
var __vue_staticRenderFns__$q = [function () {var _vm=this;var _h=_vm.$createElement;var _c=_vm._self._c||_h;return _c('div',{staticClass:"oke-qw-loading-text"},[_vm._v("\n                    Loading Questions "),_c('span',{staticClass:"oke-loader"})])}];

  /* style */
  const __vue_inject_styles__$q = function (inject) {
    if (!inject) return
    inject("data-v-5a492d6f_0", { source: ".oke-qw{margin:auto;text-align:left}.oke-qw-questionList{padding:0;margin:0}.oke-qw-main{position:relative;padding-top:24px;border-top:2px solid #f5f5f5;border-bottom:2px solid #f5f5f5}.oke-qw-empty,.oke-qw-error,.oke-qw-loading{text-align:center;padding-bottom:24px}.oke-qw-empty-button,.oke-qw-error-button,.oke-qw-loading-button{margin-top:8px}.oke-qw-empty .oke-loader,.oke-qw-error .oke-loader,.oke-qw-loading .oke-loader{margin-left:8px}.oke-qw .oke-button{display:inline-block;transition:.2s background-color ease-out;background-color:#f7f7f8;padding:.5em 24px;margin:0;border:0;border-radius:0;color:inherit;text-align:center;position:relative}.oke-qw .oke-button.oke-is-loading .oke-button-text{opacity:0}.oke-qw .oke-button.oke-is-loading::before{background-image:url(\"data:image/svg+xml,%3Csvg fill='%237c7c7c' xmlns='http://www.w3.org/2000/svg' viewBox='0 0 24 24'%3E%3Cpath d='M24 12h-2c0-5.523-4.477-10-10-10S2 6.477 2 12s4.477 10 10 10v2C5.373 24 0 18.627 0 12S5.373 0 12 0s12 5.373 12 12z'/%3E%3C/svg%3E\");background-repeat:no-repeat;display:inline-block;content:\"\";width:12px;height:12px;background-size:12px 12px;animation:oke-spin 1s infinite linear;position:absolute;top:0;bottom:0;left:0;right:0;margin:auto}.oke-qw .oke-button:focus,.oke-qw .oke-button:hover{outline:0}.oke-qw .oke-button:not(.oke-is-loading){cursor:pointer}.oke-qw .oke-button:not(.oke-is-loading):focus,.oke-qw .oke-button:not(.oke-is-loading):hover{background-color:#e5e5eb}.oke-qw .oke-is-small .oke-button{padding-top:12px;padding-bottom:12px}.oke-qw .oke-select{cursor:pointer;background-repeat:no-repeat;background-position-x:100%;background-position-y:50%;border:none;padding:0 24px 0 0;border-radius:4px;appearance:none;color:inherit;background-color:transparent;-webkit-appearance:none;background-image:url(\"data:image/svg+xml,%3Csvg fill='%23000' xmlns='http://www.w3.org/2000/svg' viewBox='0 0 24 24'%3E%3Cpath d='M7 10l5 5 5-5z'/%3E%3Cpath d='M0 0h24v24H0z' fill='none'/%3E%3C/svg%3E\");background-repeat:no-repeat}.oke-qw .oke-select:focus{outline:0;text-decoration:underline}.oke-qw .oke-select:disabled{background-color:transparent;background-image:url(\"data:image/svg+xml,%3Csvg fill='%237c7c7c' xmlns='http://www.w3.org/2000/svg' viewBox='0 0 24 24'%3E%3Cpath d='M7 10l5 5 5-5z'/%3E%3Cpath d='M0 0h24v24H0z' fill='none'/%3E%3C/svg%3E\");background-repeat:no-repeat}.oke-qw .oke-loader{background-image:url(\"data:image/svg+xml,%3Csvg fill='%237c7c7c' xmlns='http://www.w3.org/2000/svg' viewBox='0 0 24 24'%3E%3Cpath d='M24 12h-2c0-5.523-4.477-10-10-10S2 6.477 2 12s4.477 10 10 10v2C5.373 24 0 18.627 0 12S5.373 0 12 0s12 5.373 12 12z'/%3E%3C/svg%3E\");background-repeat:no-repeat;display:inline-block;content:\"\";width:12px;height:12px;background-size:12px 12px;animation:oke-spin 1s infinite linear;vertical-align:middle;user-select:none}@keyframes oke-spin{from{transform:rotate(0)}to{transform:rotate(360deg)}}", map: undefined, media: undefined });

  };
  /* scoped */
  const __vue_scope_id__$q = undefined;
  /* module identifier */
  const __vue_module_identifier__$q = undefined;
  /* functional template */
  const __vue_is_functional_template__$q = false;
  /* style inject SSR */
  
  /* style inject shadow dom */
  

  
  var QuestionsWidget = normalizeComponent(
    { render: __vue_render__$q, staticRenderFns: __vue_staticRenderFns__$q },
    __vue_inject_styles__$q,
    __vue_script__$q,
    __vue_scope_id__$q,
    __vue_is_functional_template__$q,
    __vue_module_identifier__$q,
    false,
    createInjector,
    undefined,
    undefined
  );

//
var script$r = {
  props: {
    activeTab: {
      type: String,
      required: false,
      default: 'reviews'
    },
    productId: {
      type: String,
      required: true
    },
    tabs: {
      type: Array,
      required: true
    }
  },

  data() {
    return {
      questionCount: 0,
      reviewCount: 0
    };
  },

  mounted() {
    if (this.productId) {
      Promise.all([OkendoAPI.getReviewAggregateProduct(this.productId), OkendoAPI.getQuestionAggregateProduct(this.productId)]).then(([reviewsResponse, questionResponse]) => {
        this.reviewCount = reviewsResponse.data.reviewAggregate.reviewCount || 0;
        this.questionCount = questionResponse.data.questionAggregate.approvedQuestionCount || 0;
      });
    }
  },

  methods: {
    onActiveTabChange(newTab) {
      const newTabRefs = this.$refs[newTab];

      if (newTabRefs[0]) {
        newTabRefs[0].focus();
      }

      this.$emit('tabChange', newTab);
    },

    selectFirstTab() {
      const newTab = this.tabs[0];
      this.onActiveTabChange(newTab);
    },

    selectLastTab() {
      const newTab = this.tabs[this.tabs.length - 1];
      this.onActiveTabChange(newTab);
    },

    selectNextTab(index) {
      const nextTab = this.tabs[index + 1];
      const newTab = nextTab || this.tabs[0];
      this.onActiveTabChange(newTab);
    },

    selectPreviousTab(index) {
      const prevTab = this.tabs[index - 1];
      const newTab = prevTab || this.tabs[this.tabs.length - 1];
      this.onActiveTabChange(newTab);
    }

  }
};

/* script */
const __vue_script__$r = script$r;

/* template */
var __vue_render__$r = function () {var _vm=this;var _h=_vm.$createElement;var _c=_vm._self._c||_h;return _c('div',{staticClass:"oke-rw-navBar",attrs:{"role":"tablist"}},_vm._l((_vm.tabs),function(tab,index){return _c('button',{key:tab,ref:tab,refInFor:true,staticClass:"oke-rw-navBar-item",attrs:{"aria-selected":_vm.activeTab === tab,"id":(tab + "-tab"),"aria-controls":(tab + "-panel"),"tabindex":_vm.activeTab === tab ? 0 : -1,"role":"tab","type":"button"},on:{"click":function($event){return _vm.onActiveTabChange(tab)},"keyup":[function($event){if(!$event.type.indexOf('key')&&_vm._k($event.keyCode,"right",39,$event.key,["Right","ArrowRight"])){ return null; }if('button' in $event && $event.button !== 2){ return null; }return _vm.selectNextTab(index)},function($event){if(!$event.type.indexOf('key')&&_vm._k($event.keyCode,"left",37,$event.key,["Left","ArrowLeft"])){ return null; }if('button' in $event && $event.button !== 0){ return null; }return _vm.selectPreviousTab(index)},function($event){if(!$event.type.indexOf('key')&&_vm._k($event.keyCode,"home",undefined,$event.key,undefined)){ return null; }return _vm.selectFirstTab()},function($event){if(!$event.type.indexOf('key')&&_vm._k($event.keyCode,"end",undefined,$event.key,undefined)){ return null; }return _vm.selectLastTab()}]}},[_vm._v("\n        "+_vm._s(tab)+"\n        "),(tab === 'reviews' && _vm.reviewCount)?_c('span',{staticClass:"oke-rw-navBar-item-count"},[_vm._v(_vm._s(_vm.reviewCount))]):_vm._e(),_vm._v(" "),(tab === 'questions' && _vm.questionCount)?_c('span',{staticClass:"oke-rw-navBar-item-count"},[_vm._v(_vm._s(_vm.questionCount))]):_vm._e()])}),0)};
var __vue_staticRenderFns__$r = [];

  /* style */
  const __vue_inject_styles__$r = function (inject) {
    if (!inject) return
    inject("data-v-229d67ee_0", { source: ".okeReviews .oke-rw-navBar{margin-bottom:16px;text-align:left}.okeReviews .oke-rw-navBar-item{outline:0;margin-right:24px;padding-bottom:4px;font-weight:600;background:0 0;border:0;color:inherit;cursor:pointer;display:inline-block;height:auto;padding:0;text-transform:capitalize}.okeReviews .oke-rw-navBar-item a{color:inherit}.okeReviews .oke-rw-navBar-item[aria-selected=true]{border-bottom:3px solid #dbdde4}.okeReviews .oke-rw-navBar-item-count{display:inline-block}.okeReviews .oke-rw-navBar-item-count::before{content:\"(\"}.okeReviews .oke-rw-navBar-item-count::after{content:\")\"}", map: undefined, media: undefined });

  };
  /* scoped */
  const __vue_scope_id__$r = undefined;
  /* module identifier */
  const __vue_module_identifier__$r = undefined;
  /* functional template */
  const __vue_is_functional_template__$r = false;
  /* style inject SSR */
  
  /* style inject shadow dom */
  

  
  var NavBar = normalizeComponent(
    { render: __vue_render__$r, staticRenderFns: __vue_staticRenderFns__$r },
    __vue_inject_styles__$r,
    __vue_script__$r,
    __vue_scope_id__$r,
    __vue_is_functional_template__$r,
    __vue_module_identifier__$r,
    false,
    createInjector,
    undefined,
    undefined
  );

//
var script$s = {
  props: {
    productId: {
      type: String,
      required: true
    }
  },
  components: {
    QuestionsWidget,
    ReviewsWidget,
    NavBar
  },

  data() {
    return {
      activeTab: 'reviews',
      questionsTabInitialised: false,
      tabs: ['reviews', 'questions'],
      widgetSize: 'small'
    };
  },

  mounted() {
    this.addResizeListener();
  },

  destroyed() {
    this.removeResizeListener();
  },

  methods: {
    onTabChange(newTab) {
      if (newTab === 'questions') {
        this.questionsTabInitialised = true;
      }

      this.activeTab = newTab;
    },

    addResizeListener() {
      this.setWidgetSize();
      window.addEventListener('resize', this.setWidgetSize, {
        passive: true
      });
    },

    removeResizeListener() {
      window.removeEventListener('resize', this.setWidgetSize);
    },

    setWidgetSize() {
      const widgetWidth = this.$refs.widget && this.$refs.widget.clientWidth;

      if (widgetWidth > 960) {
        this.widgetSize = 'large';
      } else if (widgetWidth > 640) {
        this.widgetSize = 'medium';
      } else {
        this.widgetSize = 'small';
      }
    }

  }
};

/* script */
const __vue_script__$s = script$s;

/* template */
var __vue_render__$s = function () {var _vm=this;var _h=_vm.$createElement;var _c=_vm._self._c||_h;return _c('div',{staticClass:"okeReviews",attrs:{"data-oke-reviews-version":""}},[_c('div',{ref:"widget",staticClass:"oke-cw"},[_c('nav-bar',{attrs:{"tabs":_vm.tabs,"activeTab":_vm.activeTab,"productId":_vm.productId},on:{"tabChange":function($event){return _vm.onTabChange($event)}}}),_vm._v(" "),_vm._l((_vm.tabs),function(tab){return _c('div',{key:tab},[_c('div',{staticClass:"oke-cw-panel",attrs:{"hidden":_vm.activeTab !== tab,"id":(tab + "-panel"),"aria-labelledby":(tab + "-tab"),"tabindex":"0","role":"tabpanel"}},[(tab === 'reviews')?_c('reviews-widget',{attrs:{"productId":_vm.productId,"sizeOverride":_vm.widgetSize}}):_vm._e(),_vm._v(" "),(_vm.questionsTabInitialised && tab === 'questions')?_c('questions-widget',{attrs:{"productId":_vm.productId,"sizeOverride":_vm.widgetSize}}):_vm._e()],1)])})],2)])};
var __vue_staticRenderFns__$s = [];

  /* style */
  const __vue_inject_styles__$s = function (inject) {
    if (!inject) return
    inject("data-v-15bb2782_0", { source: ".oke-cw-panel{outline:0}.oke-cw-panel[hidden]{display:none}", map: undefined, media: undefined });

  };
  /* scoped */
  const __vue_scope_id__$s = undefined;
  /* module identifier */
  const __vue_module_identifier__$s = undefined;
  /* functional template */
  const __vue_is_functional_template__$s = false;
  /* style inject SSR */
  
  /* style inject shadow dom */
  

  
  var OkendoCombinedWidget = normalizeComponent(
    { render: __vue_render__$s, staticRenderFns: __vue_staticRenderFns__$s },
    __vue_inject_styles__$s,
    __vue_script__$s,
    __vue_scope_id__$s,
    __vue_is_functional_template__$s,
    __vue_module_identifier__$s,
    false,
    createInjector,
    undefined,
    undefined
  );



var components = /*#__PURE__*/Object.freeze({
    __proto__: null,
    OkendoStarRating: OkendoStarRating,
    OkendoReviewsWidget: ReviewsWidget,
    OkendoQuestionsWidget: QuestionsWidget,
    OkendoCombinedWidget: OkendoCombinedWidget
});

// Generated by: https://www.npmjs.com/package/vue-sfc-rollup

function install(Vue) {
  if (install.installed) return;
  install.installed = true;
  Object.keys(components).forEach(componentName => {
    Vue.component(componentName, components[componentName]);
  });
} // Create module definition for Vue.use()


const plugin = {
  install
}; // To auto-install when vue is found

/* global window global */

let GlobalVue = null;

if (typeof window !== 'undefined') {
  GlobalVue = window.Vue;
} else if (typeof global !== 'undefined') {
  GlobalVue = global.Vue;
}

if (GlobalVue) {
  GlobalVue.use(plugin);
} // Default export is library as a whole, registered via Vue.use()
function configureOkendo(config) {
  if (config.subscriberId) {
    OkendoAPI.setSubscriberId(config.subscriberId);
  }
}

export default plugin;
export { OkendoCombinedWidget, QuestionsWidget as OkendoQuestionsWidget, ReviewsWidget as OkendoReviewsWidget, OkendoStarRating, configureOkendo };
